{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6ffd7ccd",
   "metadata": {},
   "source": [
    "## Ranks and topics: hSBM results on protein families"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "bda663e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import math\n",
    "import numpy as np\n",
    "from collections import deque\n",
    "import matplotlib.pyplot as plt\n",
    "import tqdm as tqdm\n",
    "import seaborn as sns\n",
    "from bisect import bisect_left, bisect_right\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import sys\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a7766e5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxid_to_name_path = '../taxonomy_ncbi/taxid_to_name.tsv'\n",
    "taxid_to_parent_path = '../taxonomy_ncbi/taxid_to_parent.names.ranks.tsv'\n",
    "\n",
    "tid2nm = {}\n",
    "with open(taxid_to_name_path, 'r') as f:\n",
    "\tfor line in f:\n",
    "\t\ttid, name = line.strip().split('\\t')\n",
    "\t\ttid2nm[int(tid)] = name\n",
    "                \n",
    "nm2tid = {v: k for k, v in tid2nm.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0a56dc7",
   "metadata": {},
   "source": [
    "________________"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387cc5e",
   "metadata": {},
   "source": [
    "### Count matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1c14193d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PF00002</th>\n",
       "      <th>PF00004</th>\n",
       "      <th>PF00005</th>\n",
       "      <th>PF00006</th>\n",
       "      <th>PF00008</th>\n",
       "      <th>PF00009</th>\n",
       "      <th>PF00010</th>\n",
       "      <th>PF00011</th>\n",
       "      <th>PF00012</th>\n",
       "      <th>PF00013</th>\n",
       "      <th>...</th>\n",
       "      <th>PF25791</th>\n",
       "      <th>PF25792</th>\n",
       "      <th>PF25794</th>\n",
       "      <th>PF25796</th>\n",
       "      <th>PF25799</th>\n",
       "      <th>PF25800</th>\n",
       "      <th>PF25815</th>\n",
       "      <th>PF25816</th>\n",
       "      <th>PF25818</th>\n",
       "      <th>PF25819</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>UP000000212_1234679</th>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP000000231_312153</th>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP000000233_379731</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>58</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP000000235_369723</th>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>96</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP000000238_349521</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>91</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP001268683_1456591</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP001288380_3079793</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP001289135_1332059</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP001296776_85075</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UP001296967_49447</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9098 rows × 14693 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     PF00002  PF00004  PF00005  PF00006  PF00008  PF00009  \\\n",
       "species                                                                     \n",
       "UP000000212_1234679        0        6       95        5        0        9   \n",
       "UP000000231_312153         0        8       30        3        0        7   \n",
       "UP000000233_379731         0       13       58        4        0        9   \n",
       "UP000000235_369723         0        9       96        3        0        8   \n",
       "UP000000238_349521         0       13       91        9        0        9   \n",
       "...                      ...      ...      ...      ...      ...      ...   \n",
       "UP001268683_1456591        0        0        0        0        0        0   \n",
       "UP001288380_3079793        0        0        0        0        0        0   \n",
       "UP001289135_1332059        0        0        0        0        0        0   \n",
       "UP001296776_85075          0        0        0        0        0        0   \n",
       "UP001296967_49447          0        0        0        0        0        0   \n",
       "\n",
       "                     PF00010  PF00011  PF00012  PF00013  ...  PF25791  \\\n",
       "species                                                  ...            \n",
       "UP000000212_1234679        0        0        2        2  ...        0   \n",
       "UP000000231_312153         0        0        2        1  ...        0   \n",
       "UP000000233_379731         0        2        3        1  ...        0   \n",
       "UP000000235_369723         0        1        2        2  ...        0   \n",
       "UP000000238_349521         0        2        5        1  ...        0   \n",
       "...                      ...      ...      ...      ...  ...      ...   \n",
       "UP001268683_1456591        0        0        0        0  ...        0   \n",
       "UP001288380_3079793        0        0        0        0  ...        0   \n",
       "UP001289135_1332059        0        0        0        0  ...        0   \n",
       "UP001296776_85075          0        0        0        0  ...        0   \n",
       "UP001296967_49447          0        0        0        0  ...        0   \n",
       "\n",
       "                     PF25792  PF25794  PF25796  PF25799  PF25800  PF25815  \\\n",
       "species                                                                     \n",
       "UP000000212_1234679        0        0        0        0        0        0   \n",
       "UP000000231_312153         0        0        0        0        1        0   \n",
       "UP000000233_379731         0        0        0        0        2        0   \n",
       "UP000000235_369723         0        0        0        0        0        0   \n",
       "UP000000238_349521         0        0        0        0        1        0   \n",
       "...                      ...      ...      ...      ...      ...      ...   \n",
       "UP001268683_1456591        0        0        0        0        0        0   \n",
       "UP001288380_3079793        0        0        0        0        0        0   \n",
       "UP001289135_1332059        0        0        0        0        0        0   \n",
       "UP001296776_85075          0        0        0        0        0        0   \n",
       "UP001296967_49447          0        0        0        0        0        0   \n",
       "\n",
       "                     PF25816  PF25818  PF25819  \n",
       "species                                         \n",
       "UP000000212_1234679        0        0        0  \n",
       "UP000000231_312153         0        0        0  \n",
       "UP000000233_379731         0        0        0  \n",
       "UP000000235_369723         0        0        0  \n",
       "UP000000238_349521         0        0        0  \n",
       "...                      ...      ...      ...  \n",
       "UP001268683_1456591        0        0        0  \n",
       "UP001288380_3079793        0        0        0  \n",
       "UP001289135_1332059        0        0        0  \n",
       "UP001296776_85075          0        0        0  \n",
       "UP001296967_49447          0        0        0  \n",
       "\n",
       "[9098 rows x 14693 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "countmatrix_path = '../uniprotkb_reference_proteomes/big_bacteria_countmatrix.csv'\n",
    "df_cm = pd.read_csv(countmatrix_path, index_col=0)\n",
    "df_cm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5e84d317",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PF00002</th>\n",
       "      <th>PF00004</th>\n",
       "      <th>PF00005</th>\n",
       "      <th>PF00006</th>\n",
       "      <th>PF00008</th>\n",
       "      <th>PF00009</th>\n",
       "      <th>PF00010</th>\n",
       "      <th>PF00011</th>\n",
       "      <th>PF00012</th>\n",
       "      <th>PF00013</th>\n",
       "      <th>...</th>\n",
       "      <th>PF25791</th>\n",
       "      <th>PF25792</th>\n",
       "      <th>PF25794</th>\n",
       "      <th>PF25796</th>\n",
       "      <th>PF25799</th>\n",
       "      <th>PF25800</th>\n",
       "      <th>PF25815</th>\n",
       "      <th>PF25816</th>\n",
       "      <th>PF25818</th>\n",
       "      <th>PF25819</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1234679</th>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>95</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312153</th>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379731</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>58</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369723</th>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>96</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349521</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>91</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1456591</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3079793</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1332059</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85075</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49447</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9098 rows × 14693 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         PF00002  PF00004  PF00005  PF00006  PF00008  PF00009  PF00010  \\\n",
       "1234679        0        6       95        5        0        9        0   \n",
       "312153         0        8       30        3        0        7        0   \n",
       "379731         0       13       58        4        0        9        0   \n",
       "369723         0        9       96        3        0        8        0   \n",
       "349521         0       13       91        9        0        9        0   \n",
       "...          ...      ...      ...      ...      ...      ...      ...   \n",
       "1456591        0        0        0        0        0        0        0   \n",
       "3079793        0        0        0        0        0        0        0   \n",
       "1332059        0        0        0        0        0        0        0   \n",
       "85075          0        0        0        0        0        0        0   \n",
       "49447          0        0        0        0        0        0        0   \n",
       "\n",
       "         PF00011  PF00012  PF00013  ...  PF25791  PF25792  PF25794  PF25796  \\\n",
       "1234679        0        2        2  ...        0        0        0        0   \n",
       "312153         0        2        1  ...        0        0        0        0   \n",
       "379731         2        3        1  ...        0        0        0        0   \n",
       "369723         1        2        2  ...        0        0        0        0   \n",
       "349521         2        5        1  ...        0        0        0        0   \n",
       "...          ...      ...      ...  ...      ...      ...      ...      ...   \n",
       "1456591        0        0        0  ...        0        0        0        0   \n",
       "3079793        0        0        0  ...        0        0        0        0   \n",
       "1332059        0        0        0  ...        0        0        0        0   \n",
       "85075          0        0        0  ...        0        0        0        0   \n",
       "49447          0        0        0  ...        0        0        0        0   \n",
       "\n",
       "         PF25799  PF25800  PF25815  PF25816  PF25818  PF25819  \n",
       "1234679        0        0        0        0        0        0  \n",
       "312153         0        1        0        0        0        0  \n",
       "379731         0        2        0        0        0        0  \n",
       "369723         0        0        0        0        0        0  \n",
       "349521         0        1        0        0        0        0  \n",
       "...          ...      ...      ...      ...      ...      ...  \n",
       "1456591        0        0        0        0        0        0  \n",
       "3079793        0        0        0        0        0        0  \n",
       "1332059        0        0        0        0        0        0  \n",
       "85075          0        0        0        0        0        0  \n",
       "49447          0        0        0        0        0        0  \n",
       "\n",
       "[9098 rows x 14693 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = [x.split('_')[1] for x in df_cm.index]\n",
    "df_cm.index = idx\n",
    "df_cm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42595fb6",
   "metadata": {},
   "source": [
    "_____________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e9bcc91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of connected components in the tree: 1\n",
      "Number of weakly connected components in the directed tree: 1\n",
      "Number of nodes in the tree: 2703951\n",
      "Number of edges in the tree: 2703950\n"
     ]
    }
   ],
   "source": [
    "tree = nx.Graph()\n",
    "tree_dir = nx.DiGraph()\n",
    "\n",
    "with open(taxid_to_parent_path, 'r') as f:\n",
    "    for line in f:\n",
    "        node_tid, node_rank, node_name, parent_tid, parent_rank, parent_name = line.rstrip('\\n').split('\\t')\n",
    "\n",
    "        # normalize ids (keep as strings or cast to int consistently)\n",
    "        node_tid = int(node_tid)\n",
    "        parent_tid = int(parent_tid)\n",
    "\n",
    "        # add nodes with attributes\n",
    "        tree.add_node(node_tid, name=node_name.strip(), rank=node_rank.strip())\n",
    "        tree_dir.add_node(node_tid, name=node_name.strip(), rank=node_rank.strip())\n",
    "        tree.add_node(parent_tid, name=parent_name.strip(), rank=parent_rank.strip())\n",
    "        tree_dir.add_node(parent_tid, name=parent_name.strip(), rank=parent_rank.strip())\n",
    "\n",
    "        # parent -> child\n",
    "        tree.add_edge(parent_tid, node_tid)\n",
    "        tree_dir.add_edge(parent_tid, node_tid)\n",
    "\n",
    "tree.remove_edge(1, 1)\n",
    "tree_dir.remove_edge(1, 1)\n",
    "\n",
    "print(f'Number of connected components in the tree: {nx.number_connected_components(tree)}')\n",
    "print(f'Number of weakly connected components in the directed tree: {nx.number_weakly_connected_components(tree_dir)}')\n",
    "\n",
    "print(f'Number of nodes in the tree: {tree.number_of_nodes()}')\n",
    "print(f'Number of edges in the tree: {tree.number_of_edges()}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e36248f6",
   "metadata": {},
   "source": [
    "### hSBM results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "052982ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "lvl = 0\n",
    "clusters_path = f'./results/topsbm_level_{str(lvl)}_clusters.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "830f2631",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of clusters at level 0: 48\n"
     ]
    }
   ],
   "source": [
    "clusters_df = pd.read_csv(clusters_path)\n",
    "n_clusters = clusters_df.shape[1]\n",
    "print(f'Number of clusters at level {lvl}: {n_clusters}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "093d64f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Cluster 3</th>\n",
       "      <th>Cluster 4</th>\n",
       "      <th>Cluster 5</th>\n",
       "      <th>Cluster 6</th>\n",
       "      <th>Cluster 7</th>\n",
       "      <th>Cluster 8</th>\n",
       "      <th>Cluster 9</th>\n",
       "      <th>Cluster 10</th>\n",
       "      <th>...</th>\n",
       "      <th>Cluster 39</th>\n",
       "      <th>Cluster 40</th>\n",
       "      <th>Cluster 41</th>\n",
       "      <th>Cluster 42</th>\n",
       "      <th>Cluster 43</th>\n",
       "      <th>Cluster 44</th>\n",
       "      <th>Cluster 45</th>\n",
       "      <th>Cluster 46</th>\n",
       "      <th>Cluster 47</th>\n",
       "      <th>Cluster 48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UP000001136_367737</td>\n",
       "      <td>UP000000813_176299</td>\n",
       "      <td>UP000014118_397291</td>\n",
       "      <td>UP000001115_316278</td>\n",
       "      <td>UP000016648_1115809</td>\n",
       "      <td>UP000007809_675635</td>\n",
       "      <td>UP000008871_393595</td>\n",
       "      <td>UP000000640_196162</td>\n",
       "      <td>UP000012063_1293054</td>\n",
       "      <td>UP000004263_207949</td>\n",
       "      <td>...</td>\n",
       "      <td>UP000001902_591001</td>\n",
       "      <td>UP000003195_706434</td>\n",
       "      <td>UP000005439_679936</td>\n",
       "      <td>UP000007332_1071400</td>\n",
       "      <td>UP000015503_1245471</td>\n",
       "      <td>UP000007113_682795</td>\n",
       "      <td>UP000014154_1235796</td>\n",
       "      <td>UP000005286_879305</td>\n",
       "      <td>UP000002415_381764</td>\n",
       "      <td>UP000002480_272947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UP000002495_235279</td>\n",
       "      <td>UP000002040_137722</td>\n",
       "      <td>UP000003438_411471</td>\n",
       "      <td>UP000001961_64471</td>\n",
       "      <td>UP000006657_709991</td>\n",
       "      <td>UP000004245_525370</td>\n",
       "      <td>UP000014568_421052</td>\n",
       "      <td>UP000003868_312284</td>\n",
       "      <td>UP000006427_469381</td>\n",
       "      <td>UP000006286_930169</td>\n",
       "      <td>...</td>\n",
       "      <td>UP000016640_1321779</td>\n",
       "      <td>UP000010802_1209989</td>\n",
       "      <td>UP000000323_525904</td>\n",
       "      <td>UP000007271_1185325</td>\n",
       "      <td>UP000017050_1203578</td>\n",
       "      <td>UP000006844_401053</td>\n",
       "      <td>UP000012651_997872</td>\n",
       "      <td>UP000005984_525254</td>\n",
       "      <td>UP000004793_511051</td>\n",
       "      <td>UP000007794_291272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UP000002043_638303</td>\n",
       "      <td>UP000005258_1110502</td>\n",
       "      <td>UP000014129_1235835</td>\n",
       "      <td>UP000000788_93059</td>\n",
       "      <td>UP000004295_553175</td>\n",
       "      <td>UP000001584_83332</td>\n",
       "      <td>UP000003511_1055526</td>\n",
       "      <td>UP000008229_469383</td>\n",
       "      <td>UP000006804_688269</td>\n",
       "      <td>UP000006282_385025</td>\n",
       "      <td>...</td>\n",
       "      <td>UP000007093_568816</td>\n",
       "      <td>UP000003277_742743</td>\n",
       "      <td>UP000004508_485913</td>\n",
       "      <td>UP000001285_714313</td>\n",
       "      <td>UP000002515_223283</td>\n",
       "      <td>UP000019151_861299</td>\n",
       "      <td>UP000003056_443342</td>\n",
       "      <td>UP000003422_997350</td>\n",
       "      <td>UP000006889_632518</td>\n",
       "      <td>UP000001174_177416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UP000005981_392423</td>\n",
       "      <td>UP000018524_1287292</td>\n",
       "      <td>UP000014134_1235800</td>\n",
       "      <td>UP000010379_195253</td>\n",
       "      <td>UP000004394_862515</td>\n",
       "      <td>UP000001419_233413</td>\n",
       "      <td>UP000001231_523791</td>\n",
       "      <td>UP000001116_266940</td>\n",
       "      <td>UP000006866_572479</td>\n",
       "      <td>UP000017838_1408164</td>\n",
       "      <td>...</td>\n",
       "      <td>UP000003503_888062</td>\n",
       "      <td>UP000000272_555079</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UP000001279_1069534</td>\n",
       "      <td>UP000001062_717774</td>\n",
       "      <td>UP000002432_204669</td>\n",
       "      <td>UP000005947_525256</td>\n",
       "      <td>UP000005244_796937</td>\n",
       "      <td>UP000002382_521045</td>\n",
       "      <td>UP000018554_1073362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UP000007803_563040</td>\n",
       "      <td>UP000001054_394</td>\n",
       "      <td>UP000014150_397290</td>\n",
       "      <td>UP000001420_167539</td>\n",
       "      <td>UP000006545_879243</td>\n",
       "      <td>UP000005143_1097667</td>\n",
       "      <td>UP000007477_871585</td>\n",
       "      <td>UP000008221_351607</td>\n",
       "      <td>UP000005096_584708</td>\n",
       "      <td>UP000002677_83406</td>\n",
       "      <td>...</td>\n",
       "      <td>UP000002072_519441</td>\n",
       "      <td>UP000010880_748449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UP000001259_390333</td>\n",
       "      <td>UP000000233_379731</td>\n",
       "      <td>UP000002207_240015</td>\n",
       "      <td>UP000004830_742742</td>\n",
       "      <td>UP000003280_862517</td>\n",
       "      <td>UP000007719_515635</td>\n",
       "      <td>UP000004116_1005043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UP000000432_220668</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UP000005147_883112</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UP000019050_592010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UP000005388_764291</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UP000018559_1392007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Cluster 1            Cluster 2            Cluster 3  \\\n",
       "0   UP000001136_367737   UP000000813_176299   UP000014118_397291   \n",
       "1   UP000002495_235279   UP000002040_137722   UP000003438_411471   \n",
       "2   UP000002043_638303  UP000005258_1110502  UP000014129_1235835   \n",
       "3   UP000005981_392423  UP000018524_1287292  UP000014134_1235800   \n",
       "4   UP000007803_563040      UP000001054_394   UP000014150_397290   \n",
       "..                 ...                  ...                  ...   \n",
       "73                 NaN                  NaN                  NaN   \n",
       "74                 NaN                  NaN                  NaN   \n",
       "75                 NaN                  NaN                  NaN   \n",
       "76                 NaN                  NaN                  NaN   \n",
       "77                 NaN                  NaN                  NaN   \n",
       "\n",
       "             Cluster 4            Cluster 5            Cluster 6  \\\n",
       "0   UP000001115_316278  UP000016648_1115809   UP000007809_675635   \n",
       "1    UP000001961_64471   UP000006657_709991   UP000004245_525370   \n",
       "2    UP000000788_93059   UP000004295_553175    UP000001584_83332   \n",
       "3   UP000010379_195253   UP000004394_862515   UP000001419_233413   \n",
       "4   UP000001420_167539   UP000006545_879243  UP000005143_1097667   \n",
       "..                 ...                  ...                  ...   \n",
       "73                 NaN                  NaN                  NaN   \n",
       "74                 NaN                  NaN                  NaN   \n",
       "75                 NaN                  NaN                  NaN   \n",
       "76                 NaN                  NaN                  NaN   \n",
       "77                 NaN                  NaN                  NaN   \n",
       "\n",
       "              Cluster 7           Cluster 8            Cluster 9  \\\n",
       "0    UP000008871_393595  UP000000640_196162  UP000012063_1293054   \n",
       "1    UP000014568_421052  UP000003868_312284   UP000006427_469381   \n",
       "2   UP000003511_1055526  UP000008229_469383   UP000006804_688269   \n",
       "3    UP000001231_523791  UP000001116_266940   UP000006866_572479   \n",
       "4    UP000007477_871585  UP000008221_351607   UP000005096_584708   \n",
       "..                  ...                 ...                  ...   \n",
       "73                  NaN                 NaN                  NaN   \n",
       "74                  NaN                 NaN                  NaN   \n",
       "75                  NaN                 NaN                  NaN   \n",
       "76                  NaN                 NaN                  NaN   \n",
       "77                  NaN                 NaN                  NaN   \n",
       "\n",
       "             Cluster 10  ...           Cluster 39           Cluster 40  \\\n",
       "0    UP000004263_207949  ...   UP000001902_591001   UP000003195_706434   \n",
       "1    UP000006286_930169  ...  UP000016640_1321779  UP000010802_1209989   \n",
       "2    UP000006282_385025  ...   UP000007093_568816   UP000003277_742743   \n",
       "3   UP000017838_1408164  ...   UP000003503_888062   UP000000272_555079   \n",
       "4     UP000002677_83406  ...   UP000002072_519441   UP000010880_748449   \n",
       "..                  ...  ...                  ...                  ...   \n",
       "73                  NaN  ...                  NaN                  NaN   \n",
       "74                  NaN  ...                  NaN                  NaN   \n",
       "75                  NaN  ...                  NaN                  NaN   \n",
       "76                  NaN  ...                  NaN                  NaN   \n",
       "77                  NaN  ...                  NaN                  NaN   \n",
       "\n",
       "            Cluster 41           Cluster 42           Cluster 43  \\\n",
       "0   UP000005439_679936  UP000007332_1071400  UP000015503_1245471   \n",
       "1   UP000000323_525904  UP000007271_1185325  UP000017050_1203578   \n",
       "2   UP000004508_485913   UP000001285_714313   UP000002515_223283   \n",
       "3                  NaN  UP000001279_1069534   UP000001062_717774   \n",
       "4                  NaN   UP000001259_390333   UP000000233_379731   \n",
       "..                 ...                  ...                  ...   \n",
       "73                 NaN   UP000000432_220668                  NaN   \n",
       "74                 NaN   UP000005147_883112                  NaN   \n",
       "75                 NaN   UP000019050_592010                  NaN   \n",
       "76                 NaN   UP000005388_764291                  NaN   \n",
       "77                 NaN  UP000018559_1392007                  NaN   \n",
       "\n",
       "            Cluster 44           Cluster 45          Cluster 46  \\\n",
       "0   UP000007113_682795  UP000014154_1235796  UP000005286_879305   \n",
       "1   UP000006844_401053   UP000012651_997872  UP000005984_525254   \n",
       "2   UP000019151_861299   UP000003056_443342  UP000003422_997350   \n",
       "3   UP000002432_204669   UP000005947_525256  UP000005244_796937   \n",
       "4   UP000002207_240015   UP000004830_742742  UP000003280_862517   \n",
       "..                 ...                  ...                 ...   \n",
       "73                 NaN                  NaN                 NaN   \n",
       "74                 NaN                  NaN                 NaN   \n",
       "75                 NaN                  NaN                 NaN   \n",
       "76                 NaN                  NaN                 NaN   \n",
       "77                 NaN                  NaN                 NaN   \n",
       "\n",
       "            Cluster 47           Cluster 48  \n",
       "0   UP000002415_381764   UP000002480_272947  \n",
       "1   UP000004793_511051   UP000007794_291272  \n",
       "2   UP000006889_632518   UP000001174_177416  \n",
       "3   UP000002382_521045  UP000018554_1073362  \n",
       "4   UP000007719_515635  UP000004116_1005043  \n",
       "..                 ...                  ...  \n",
       "73                 NaN                  NaN  \n",
       "74                 NaN                  NaN  \n",
       "75                 NaN                  NaN  \n",
       "76                 NaN                  NaN  \n",
       "77                 NaN                  NaN  \n",
       "\n",
       "[78 rows x 48 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b941a9f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster 1</th>\n",
       "      <th>Cluster 2</th>\n",
       "      <th>Cluster 3</th>\n",
       "      <th>Cluster 4</th>\n",
       "      <th>Cluster 5</th>\n",
       "      <th>Cluster 6</th>\n",
       "      <th>Cluster 7</th>\n",
       "      <th>Cluster 8</th>\n",
       "      <th>Cluster 9</th>\n",
       "      <th>Cluster 10</th>\n",
       "      <th>...</th>\n",
       "      <th>Cluster 39</th>\n",
       "      <th>Cluster 40</th>\n",
       "      <th>Cluster 41</th>\n",
       "      <th>Cluster 42</th>\n",
       "      <th>Cluster 43</th>\n",
       "      <th>Cluster 44</th>\n",
       "      <th>Cluster 45</th>\n",
       "      <th>Cluster 46</th>\n",
       "      <th>Cluster 47</th>\n",
       "      <th>Cluster 48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>367737.0</td>\n",
       "      <td>176299.0</td>\n",
       "      <td>397291.0</td>\n",
       "      <td>316278.0</td>\n",
       "      <td>1115809.0</td>\n",
       "      <td>675635.0</td>\n",
       "      <td>393595.0</td>\n",
       "      <td>196162.0</td>\n",
       "      <td>1293054.0</td>\n",
       "      <td>207949.0</td>\n",
       "      <td>...</td>\n",
       "      <td>591001.0</td>\n",
       "      <td>706434.0</td>\n",
       "      <td>679936.0</td>\n",
       "      <td>1071400</td>\n",
       "      <td>1245471.0</td>\n",
       "      <td>682795.0</td>\n",
       "      <td>1235796.0</td>\n",
       "      <td>879305.0</td>\n",
       "      <td>381764.0</td>\n",
       "      <td>272947.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>235279.0</td>\n",
       "      <td>137722.0</td>\n",
       "      <td>411471.0</td>\n",
       "      <td>64471.0</td>\n",
       "      <td>709991.0</td>\n",
       "      <td>525370.0</td>\n",
       "      <td>421052.0</td>\n",
       "      <td>312284.0</td>\n",
       "      <td>469381.0</td>\n",
       "      <td>930169.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1321779.0</td>\n",
       "      <td>1209989.0</td>\n",
       "      <td>525904.0</td>\n",
       "      <td>1185325</td>\n",
       "      <td>1203578.0</td>\n",
       "      <td>401053.0</td>\n",
       "      <td>997872.0</td>\n",
       "      <td>525254.0</td>\n",
       "      <td>511051.0</td>\n",
       "      <td>291272.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>638303.0</td>\n",
       "      <td>1110502.0</td>\n",
       "      <td>1235835.0</td>\n",
       "      <td>93059.0</td>\n",
       "      <td>553175.0</td>\n",
       "      <td>83332.0</td>\n",
       "      <td>1055526.0</td>\n",
       "      <td>469383.0</td>\n",
       "      <td>688269.0</td>\n",
       "      <td>385025.0</td>\n",
       "      <td>...</td>\n",
       "      <td>568816.0</td>\n",
       "      <td>742743.0</td>\n",
       "      <td>485913.0</td>\n",
       "      <td>714313</td>\n",
       "      <td>223283.0</td>\n",
       "      <td>861299.0</td>\n",
       "      <td>443342.0</td>\n",
       "      <td>997350.0</td>\n",
       "      <td>632518.0</td>\n",
       "      <td>177416.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>392423.0</td>\n",
       "      <td>1287292.0</td>\n",
       "      <td>1235800.0</td>\n",
       "      <td>195253.0</td>\n",
       "      <td>862515.0</td>\n",
       "      <td>233413.0</td>\n",
       "      <td>523791.0</td>\n",
       "      <td>266940.0</td>\n",
       "      <td>572479.0</td>\n",
       "      <td>1408164.0</td>\n",
       "      <td>...</td>\n",
       "      <td>888062.0</td>\n",
       "      <td>555079.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1069534</td>\n",
       "      <td>717774.0</td>\n",
       "      <td>204669.0</td>\n",
       "      <td>525256.0</td>\n",
       "      <td>796937.0</td>\n",
       "      <td>521045.0</td>\n",
       "      <td>1073362.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>563040.0</td>\n",
       "      <td>394.0</td>\n",
       "      <td>397290.0</td>\n",
       "      <td>167539.0</td>\n",
       "      <td>879243.0</td>\n",
       "      <td>1097667.0</td>\n",
       "      <td>871585.0</td>\n",
       "      <td>351607.0</td>\n",
       "      <td>584708.0</td>\n",
       "      <td>83406.0</td>\n",
       "      <td>...</td>\n",
       "      <td>519441.0</td>\n",
       "      <td>748449.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>390333</td>\n",
       "      <td>379731.0</td>\n",
       "      <td>240015.0</td>\n",
       "      <td>742742.0</td>\n",
       "      <td>862517.0</td>\n",
       "      <td>515635.0</td>\n",
       "      <td>1005043.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>220668</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>883112</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>592010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>764291</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1392007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Cluster 1  Cluster 2  Cluster 3  Cluster 4  Cluster 5  Cluster 6  \\\n",
       "0    367737.0   176299.0   397291.0   316278.0  1115809.0   675635.0   \n",
       "1    235279.0   137722.0   411471.0    64471.0   709991.0   525370.0   \n",
       "2    638303.0  1110502.0  1235835.0    93059.0   553175.0    83332.0   \n",
       "3    392423.0  1287292.0  1235800.0   195253.0   862515.0   233413.0   \n",
       "4    563040.0      394.0   397290.0   167539.0   879243.0  1097667.0   \n",
       "..        ...        ...        ...        ...        ...        ...   \n",
       "73        NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "74        NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "75        NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "76        NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "77        NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "\n",
       "    Cluster 7  Cluster 8  Cluster 9  Cluster 10  ...  Cluster 39  Cluster 40  \\\n",
       "0    393595.0   196162.0  1293054.0    207949.0  ...    591001.0    706434.0   \n",
       "1    421052.0   312284.0   469381.0    930169.0  ...   1321779.0   1209989.0   \n",
       "2   1055526.0   469383.0   688269.0    385025.0  ...    568816.0    742743.0   \n",
       "3    523791.0   266940.0   572479.0   1408164.0  ...    888062.0    555079.0   \n",
       "4    871585.0   351607.0   584708.0     83406.0  ...    519441.0    748449.0   \n",
       "..        ...        ...        ...         ...  ...         ...         ...   \n",
       "73        NaN        NaN        NaN         NaN  ...         NaN         NaN   \n",
       "74        NaN        NaN        NaN         NaN  ...         NaN         NaN   \n",
       "75        NaN        NaN        NaN         NaN  ...         NaN         NaN   \n",
       "76        NaN        NaN        NaN         NaN  ...         NaN         NaN   \n",
       "77        NaN        NaN        NaN         NaN  ...         NaN         NaN   \n",
       "\n",
       "    Cluster 41  Cluster 42  Cluster 43  Cluster 44  Cluster 45  Cluster 46  \\\n",
       "0     679936.0     1071400   1245471.0    682795.0   1235796.0    879305.0   \n",
       "1     525904.0     1185325   1203578.0    401053.0    997872.0    525254.0   \n",
       "2     485913.0      714313    223283.0    861299.0    443342.0    997350.0   \n",
       "3          NaN     1069534    717774.0    204669.0    525256.0    796937.0   \n",
       "4          NaN      390333    379731.0    240015.0    742742.0    862517.0   \n",
       "..         ...         ...         ...         ...         ...         ...   \n",
       "73         NaN      220668         NaN         NaN         NaN         NaN   \n",
       "74         NaN      883112         NaN         NaN         NaN         NaN   \n",
       "75         NaN      592010         NaN         NaN         NaN         NaN   \n",
       "76         NaN      764291         NaN         NaN         NaN         NaN   \n",
       "77         NaN     1392007         NaN         NaN         NaN         NaN   \n",
       "\n",
       "    Cluster 47  Cluster 48  \n",
       "0     381764.0    272947.0  \n",
       "1     511051.0    291272.0  \n",
       "2     632518.0    177416.0  \n",
       "3     521045.0   1073362.0  \n",
       "4     515635.0   1005043.0  \n",
       "..         ...         ...  \n",
       "73         NaN         NaN  \n",
       "74         NaN         NaN  \n",
       "75         NaN         NaN  \n",
       "76         NaN         NaN  \n",
       "77         NaN         NaN  \n",
       "\n",
       "[78 rows x 48 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for col in clusters_df.columns:\n",
    "    # populate cluster with taxid\n",
    "    clusters_df[col] = clusters_df[col].apply(lambda x: int(x.split('_')[1]) if pd.notnull(x) else x)\n",
    "    \n",
    "clusters_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ed93a73b",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters = {}\n",
    "for i in range(n_clusters):\n",
    "    clusters[i] = clusters_df.iloc[:, i].dropna().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d30a7d58",
   "metadata": {},
   "source": [
    "________________________________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "b6d15ccc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing cluster 1/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 3379134. Computing descendants...\n",
      "\tCluster size: 42. MRCA descendants size: 278020. Intersection size: 41.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 2/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 28211. Computing descendants...\n",
      "\tCluster size: 31. MRCA descendants size: 68707. Intersection size: 31.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 3/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1239. Computing descendants...\n",
      "\tCluster size: 40. MRCA descendants size: 90689. Intersection size: 40.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 4/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 3028117. Computing descendants...\n",
      "\tCluster size: 22. MRCA descendants size: 31831. Intersection size: 22.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 5/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 976. Computing descendants...\n",
      "\tCluster size: 52. MRCA descendants size: 23910. Intersection size: 52.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 6/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 201174. Computing descendants...\n",
      "\tCluster size: 35. MRCA descendants size: 93547. Intersection size: 34.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 7/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1224. Computing descendants...\n",
      "\tCluster size: 13. MRCA descendants size: 234365. Intersection size: 13.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 8/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 39. MRCA descendants size: 569721. Intersection size: 39.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 9/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 13. MRCA descendants size: 569721. Intersection size: 13.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 10/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1224. Computing descendants...\n",
      "\tCluster size: 20. MRCA descendants size: 234365. Intersection size: 20.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 11/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 186801. Computing descendants...\n",
      "\tCluster size: 13. MRCA descendants size: 11445. Intersection size: 13.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 12/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1224. Computing descendants...\n",
      "\tCluster size: 46. MRCA descendants size: 234365. Intersection size: 46.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 13/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1224. Computing descendants...\n",
      "\tCluster size: 34. MRCA descendants size: 234365. Intersection size: 34.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 14/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1236. Computing descendants...\n",
      "\tCluster size: 40. MRCA descendants size: 134380. Intersection size: 39.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 15/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1385. Computing descendants...\n",
      "\tCluster size: 12. MRCA descendants size: 61959. Intersection size: 12.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 16/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1760. Computing descendants...\n",
      "\tCluster size: 48. MRCA descendants size: 91371. Intersection size: 47.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 17/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 68336. Computing descendants...\n",
      "\tCluster size: 72. MRCA descendants size: 24583. Intersection size: 72.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 18/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1239. Computing descendants...\n",
      "\tCluster size: 42. MRCA descendants size: 90689. Intersection size: 39.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 19/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 29. MRCA descendants size: 569721. Intersection size: 29.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 20/48 ...\n",
      "\tComputing MRCA...\n",
      "[mrca] Skipping 1 node(s) not in the tree.\n",
      "\tMRCA found: 80840. Computing descendants...\n",
      "\tCluster size: 24. MRCA descendants size: 23352. Intersection size: 23.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 21/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 3028117. Computing descendants...\n",
      "\tCluster size: 39. MRCA descendants size: 31831. Intersection size: 39.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 22/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2818505. Computing descendants...\n",
      "\tCluster size: 4. MRCA descendants size: 932. Intersection size: 4.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 23/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 28211. Computing descendants...\n",
      "\tCluster size: 45. MRCA descendants size: 68707. Intersection size: 45.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 24/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1783257. Computing descendants...\n",
      "\tCluster size: 10. MRCA descendants size: 3924. Intersection size: 10.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 25/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 544448. Computing descendants...\n",
      "\tCluster size: 36. MRCA descendants size: 3928. Intersection size: 35.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 26/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 3379134. Computing descendants...\n",
      "\tCluster size: 18. MRCA descendants size: 278020. Intersection size: 17.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 27/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 11. MRCA descendants size: 569721. Intersection size: 11.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 28/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 48. MRCA descendants size: 569721. Intersection size: 47.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 29/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 3379134. Computing descendants...\n",
      "\tCluster size: 14. MRCA descendants size: 278020. Intersection size: 14.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 30/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 3379134. Computing descendants...\n",
      "\tCluster size: 7. MRCA descendants size: 278020. Intersection size: 7.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 31/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1783272. Computing descendants...\n",
      "\tCluster size: 35. MRCA descendants size: 223458. Intersection size: 35.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 32/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 28211. Computing descendants...\n",
      "\tCluster size: 27. MRCA descendants size: 68707. Intersection size: 26.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 33/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 203692. Computing descendants...\n",
      "\tCluster size: 20. MRCA descendants size: 3487. Intersection size: 20.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 34/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 91347. Computing descendants...\n",
      "\tCluster size: 40. MRCA descendants size: 39072. Intersection size: 32.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 35/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1783272. Computing descendants...\n",
      "\tCluster size: 32. MRCA descendants size: 223458. Intersection size: 30.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 36/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1760. Computing descendants...\n",
      "\tCluster size: 57. MRCA descendants size: 91371. Intersection size: 57.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 37/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 15. MRCA descendants size: 569721. Intersection size: 15.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 38/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 28211. Computing descendants...\n",
      "\tCluster size: 27. MRCA descendants size: 68707. Intersection size: 27.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 39/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 19. MRCA descendants size: 569721. Intersection size: 19.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 40/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1239. Computing descendants...\n",
      "\tCluster size: 16. MRCA descendants size: 90689. Intersection size: 16.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 41/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1783272. Computing descendants...\n",
      "\tCluster size: 3. MRCA descendants size: 223458. Intersection size: 3.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 42/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1783272. Computing descendants...\n",
      "\tCluster size: 78. MRCA descendants size: 223458. Intersection size: 75.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 43/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1224. Computing descendants...\n",
      "\tCluster size: 12. MRCA descendants size: 234365. Intersection size: 11.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 44/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 3379134. Computing descendants...\n",
      "\tCluster size: 8. MRCA descendants size: 278020. Intersection size: 8.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 45/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 21. MRCA descendants size: 569721. Intersection size: 21.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 46/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 1783272. Computing descendants...\n",
      "\tCluster size: 28. MRCA descendants size: 223458. Intersection size: 28.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 47/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 2. Computing descendants...\n",
      "\tCluster size: 11. MRCA descendants size: 569721. Intersection size: 11.\n",
      "\tComputing precision, recall, F1...\n",
      "Processing cluster 48/48 ...\n",
      "\tComputing MRCA...\n",
      "\tMRCA found: 3379134. Computing descendants...\n",
      "\tCluster size: 54. MRCA descendants size: 278020. Intersection size: 50.\n",
      "\tComputing precision, recall, F1...\n"
     ]
    }
   ],
   "source": [
    "def mrca(tree_dir: nx.DiGraph, nodes, *, verbose=True):\n",
    "    # --- normalize labels like 795666.0 -> 795666\n",
    "    def _normalize(u):\n",
    "        if isinstance(u, float) and u.is_integer():\n",
    "            return int(u)\n",
    "        return u\n",
    "\n",
    "    nodes = [ _normalize(u) for u in nodes ]\n",
    "    if not nodes:\n",
    "        return None\n",
    "\n",
    "    # Build parent map + find root (assumes arborescence)\n",
    "    parent = {}\n",
    "    root = None\n",
    "    for v in tree_dir.nodes:\n",
    "        preds = list(tree_dir.predecessors(v))\n",
    "        if len(preds) > 1:\n",
    "            raise ValueError(f\"Graph is not a tree (node {v} has >1 parent)\")\n",
    "        parent[v] = preds[0] if preds else None\n",
    "        if not preds:\n",
    "            root = v\n",
    "    if root is None:\n",
    "        raise ValueError(\"No root found (cycle or no in-degree-0 node)\")\n",
    "\n",
    "    # Depth from root (BFS)\n",
    "    depth = {root: 0}\n",
    "    q = deque([root])\n",
    "    while q:\n",
    "        a = q.popleft()\n",
    "        for b in tree_dir.successors(a):\n",
    "            depth[b] = depth[a] + 1\n",
    "            q.append(b)\n",
    "\n",
    "    # Keep only nodes present in this tree\n",
    "    present = [u for u in nodes if u in parent]\n",
    "    skipped = len(nodes) - len(present)\n",
    "    if verbose and skipped:\n",
    "        print(f\"[mrca] Skipping {skipped} node(s) not in the tree.\")\n",
    "\n",
    "    if len(present) < 2:\n",
    "        return None\n",
    "\n",
    "    # Ancestor set (including self); empty if node somehow missing\n",
    "    def anc_set(u):\n",
    "        s = set()\n",
    "        while u is not None and u in parent:\n",
    "            s.add(u)\n",
    "            u = parent[u]\n",
    "        return s\n",
    "\n",
    "    # Intersect ancestor sets\n",
    "    common = anc_set(present[0])\n",
    "    for u in present[1:]:\n",
    "        common &= anc_set(u)\n",
    "        if not common:\n",
    "            return None  # different components or no shared ancestor in this tree\n",
    "\n",
    "    # Deepest common ancestor\n",
    "    return max(common, key=lambda x: depth.get(x, -1))\n",
    "\n",
    "def tip_descendants(tree_dir, node):\n",
    "    tips = set()\n",
    "    q = deque([node])\n",
    "    while q:\n",
    "        u = q.popleft()\n",
    "        children = list(tree_dir.successors(u))\n",
    "        if not children:\n",
    "            tips.add(u)\n",
    "        else:\n",
    "            q.extend(children)\n",
    "    return tips\n",
    "\n",
    "def cluster_clade_stats(tree_dir, C):\n",
    "\t\n",
    "\tC = set(C)\n",
    "\t\n",
    "\tprint(f'\\tComputing MRCA...')\n",
    "\tr = mrca(tree_dir, C)\n",
    "    \n",
    "\tif r is None:\n",
    "\t\tprint(f'\\tNo MRCA found!')\n",
    "\t\treturn dict(size=len(C), precision=0.0, recall=0.0, F1=0.0, clade_like=False)\n",
    "\n",
    "\tprint(f'\\tMRCA found: {r}. Computing descendants...')\n",
    "\tD = tip_descendants(tree_dir, r)\n",
    "\tinter = len(C & D)\n",
    "\tprint(f'\\tCluster size: {len(C)}. MRCA descendants size: {len(D)}. Intersection size: {inter}.')\n",
    "\t\n",
    "\tif inter == 0:\n",
    "\t\tprint(f'\\tNo intersection between cluster and descendants of MRCA!')\n",
    "\t\treturn dict(size=len(C), precision=0.0, recall=0.0, F1=0.0, clade_like=False)\n",
    "\t\n",
    "\tprint(f'\\tComputing precision, recall, F1...')\n",
    "\tprecision = inter / len(D)\n",
    "\trecall = inter / len(C)\n",
    "\tF1 = 2 * precision * recall / (precision + recall) if (precision + recall) else 0.0\n",
    "\t\n",
    "\treturn dict(size=len(C), precision=precision, recall=recall, F1=F1,\n",
    "                clade_like=(precision == 1.0 and recall == 1.0), mrca=r, mrca_size=len(D))\n",
    "\n",
    "\n",
    "cluster_stats = {}\n",
    "for i in range(n_clusters):\n",
    "\tprint(f'Processing cluster {i+1}/{n_clusters} ...')\n",
    "\tcluster_stats[i] = cluster_clade_stats(tree_dir, clusters[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2871db5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_stats_df = pd.DataFrame.from_dict(cluster_stats, orient='index')\n",
    "cluster_stats_df\n",
    "cluster_stats_df['mrca_name'] = cluster_stats_df['mrca'].apply(lambda x: tid2nm[x] if x in tid2nm else 'unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c6000bb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>size</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>F1</th>\n",
       "      <th>clade_like</th>\n",
       "      <th>mrca</th>\n",
       "      <th>mrca_size</th>\n",
       "      <th>mrca_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>42</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.976190</td>\n",
       "      <td>0.000295</td>\n",
       "      <td>False</td>\n",
       "      <td>3379134</td>\n",
       "      <td>278020</td>\n",
       "      <td>Pseudomonadati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>0.000451</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000902</td>\n",
       "      <td>False</td>\n",
       "      <td>28211</td>\n",
       "      <td>68707</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000882</td>\n",
       "      <td>False</td>\n",
       "      <td>1239</td>\n",
       "      <td>90689</td>\n",
       "      <td>Bacillota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>0.000691</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001381</td>\n",
       "      <td>False</td>\n",
       "      <td>3028117</td>\n",
       "      <td>31831</td>\n",
       "      <td>Cyanophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>52</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.004340</td>\n",
       "      <td>False</td>\n",
       "      <td>976</td>\n",
       "      <td>23910</td>\n",
       "      <td>Bacteroidota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>35</td>\n",
       "      <td>0.000363</td>\n",
       "      <td>0.971429</td>\n",
       "      <td>0.000727</td>\n",
       "      <td>False</td>\n",
       "      <td>201174</td>\n",
       "      <td>93547</td>\n",
       "      <td>Actinomycetota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>13</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>False</td>\n",
       "      <td>1224</td>\n",
       "      <td>234365</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>39</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>13</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>20</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000171</td>\n",
       "      <td>False</td>\n",
       "      <td>1224</td>\n",
       "      <td>234365</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>13</td>\n",
       "      <td>0.001136</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.002269</td>\n",
       "      <td>False</td>\n",
       "      <td>186801</td>\n",
       "      <td>11445</td>\n",
       "      <td>Clostridia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>46</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000392</td>\n",
       "      <td>False</td>\n",
       "      <td>1224</td>\n",
       "      <td>234365</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>34</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>False</td>\n",
       "      <td>1224</td>\n",
       "      <td>234365</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>40</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.975000</td>\n",
       "      <td>0.000580</td>\n",
       "      <td>False</td>\n",
       "      <td>1236</td>\n",
       "      <td>134380</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>12</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000387</td>\n",
       "      <td>False</td>\n",
       "      <td>1385</td>\n",
       "      <td>61959</td>\n",
       "      <td>Bacillales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>48</td>\n",
       "      <td>0.000514</td>\n",
       "      <td>0.979167</td>\n",
       "      <td>0.001028</td>\n",
       "      <td>False</td>\n",
       "      <td>1760</td>\n",
       "      <td>91371</td>\n",
       "      <td>Actinomycetes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>72</td>\n",
       "      <td>0.002929</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.005841</td>\n",
       "      <td>False</td>\n",
       "      <td>68336</td>\n",
       "      <td>24583</td>\n",
       "      <td>Bacteroidota/Chlorobiota group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>42</td>\n",
       "      <td>0.000430</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>False</td>\n",
       "      <td>1239</td>\n",
       "      <td>90689</td>\n",
       "      <td>Bacillota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>29</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>24</td>\n",
       "      <td>0.000985</td>\n",
       "      <td>0.958333</td>\n",
       "      <td>0.001968</td>\n",
       "      <td>False</td>\n",
       "      <td>80840</td>\n",
       "      <td>23352</td>\n",
       "      <td>Burkholderiales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>39</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.002447</td>\n",
       "      <td>False</td>\n",
       "      <td>3028117</td>\n",
       "      <td>31831</td>\n",
       "      <td>Cyanophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>4</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.008547</td>\n",
       "      <td>False</td>\n",
       "      <td>2818505</td>\n",
       "      <td>932</td>\n",
       "      <td>Myxococcota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>45</td>\n",
       "      <td>0.000655</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001309</td>\n",
       "      <td>False</td>\n",
       "      <td>28211</td>\n",
       "      <td>68707</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>10</td>\n",
       "      <td>0.002548</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.005084</td>\n",
       "      <td>False</td>\n",
       "      <td>1783257</td>\n",
       "      <td>3924</td>\n",
       "      <td>PVC group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>36</td>\n",
       "      <td>0.008910</td>\n",
       "      <td>0.972222</td>\n",
       "      <td>0.017659</td>\n",
       "      <td>False</td>\n",
       "      <td>544448</td>\n",
       "      <td>3928</td>\n",
       "      <td>Mycoplasmatota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>18</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>False</td>\n",
       "      <td>3379134</td>\n",
       "      <td>278020</td>\n",
       "      <td>Pseudomonadati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>11</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>48</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.979167</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>14</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>False</td>\n",
       "      <td>3379134</td>\n",
       "      <td>278020</td>\n",
       "      <td>Pseudomonadati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>7</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>False</td>\n",
       "      <td>3379134</td>\n",
       "      <td>278020</td>\n",
       "      <td>Pseudomonadati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>35</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000313</td>\n",
       "      <td>False</td>\n",
       "      <td>1783272</td>\n",
       "      <td>223458</td>\n",
       "      <td>Bacillati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>27</td>\n",
       "      <td>0.000378</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>False</td>\n",
       "      <td>28211</td>\n",
       "      <td>68707</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>20</td>\n",
       "      <td>0.005736</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.011406</td>\n",
       "      <td>False</td>\n",
       "      <td>203692</td>\n",
       "      <td>3487</td>\n",
       "      <td>Spirochaetia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>40</td>\n",
       "      <td>0.000819</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.001636</td>\n",
       "      <td>False</td>\n",
       "      <td>91347</td>\n",
       "      <td>39072</td>\n",
       "      <td>Enterobacterales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>32</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.937500</td>\n",
       "      <td>0.000268</td>\n",
       "      <td>False</td>\n",
       "      <td>1783272</td>\n",
       "      <td>223458</td>\n",
       "      <td>Bacillati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>57</td>\n",
       "      <td>0.000624</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001247</td>\n",
       "      <td>False</td>\n",
       "      <td>1760</td>\n",
       "      <td>91371</td>\n",
       "      <td>Actinomycetes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>15</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>27</td>\n",
       "      <td>0.000393</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000786</td>\n",
       "      <td>False</td>\n",
       "      <td>28211</td>\n",
       "      <td>68707</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>19</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>16</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>False</td>\n",
       "      <td>1239</td>\n",
       "      <td>90689</td>\n",
       "      <td>Bacillota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>3</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>False</td>\n",
       "      <td>1783272</td>\n",
       "      <td>223458</td>\n",
       "      <td>Bacillati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>78</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>0.961538</td>\n",
       "      <td>0.000671</td>\n",
       "      <td>False</td>\n",
       "      <td>1783272</td>\n",
       "      <td>223458</td>\n",
       "      <td>Bacillati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>12</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>False</td>\n",
       "      <td>1224</td>\n",
       "      <td>234365</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>8</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>False</td>\n",
       "      <td>3379134</td>\n",
       "      <td>278020</td>\n",
       "      <td>Pseudomonadati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>21</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>28</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>False</td>\n",
       "      <td>1783272</td>\n",
       "      <td>223458</td>\n",
       "      <td>Bacillati</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>11</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>569721</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>54</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.925926</td>\n",
       "      <td>0.000360</td>\n",
       "      <td>False</td>\n",
       "      <td>3379134</td>\n",
       "      <td>278020</td>\n",
       "      <td>Pseudomonadati</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    size  precision    recall        F1  clade_like     mrca  mrca_size  \\\n",
       "0     42   0.000147  0.976190  0.000295       False  3379134     278020   \n",
       "1     31   0.000451  1.000000  0.000902       False    28211      68707   \n",
       "2     40   0.000441  1.000000  0.000882       False     1239      90689   \n",
       "3     22   0.000691  1.000000  0.001381       False  3028117      31831   \n",
       "4     52   0.002175  1.000000  0.004340       False      976      23910   \n",
       "5     35   0.000363  0.971429  0.000727       False   201174      93547   \n",
       "6     13   0.000055  1.000000  0.000111       False     1224     234365   \n",
       "7     39   0.000068  1.000000  0.000137       False        2     569721   \n",
       "8     13   0.000023  1.000000  0.000046       False        2     569721   \n",
       "9     20   0.000085  1.000000  0.000171       False     1224     234365   \n",
       "10    13   0.001136  1.000000  0.002269       False   186801      11445   \n",
       "11    46   0.000196  1.000000  0.000392       False     1224     234365   \n",
       "12    34   0.000145  1.000000  0.000290       False     1224     234365   \n",
       "13    40   0.000290  0.975000  0.000580       False     1236     134380   \n",
       "14    12   0.000194  1.000000  0.000387       False     1385      61959   \n",
       "15    48   0.000514  0.979167  0.001028       False     1760      91371   \n",
       "16    72   0.002929  1.000000  0.005841       False    68336      24583   \n",
       "17    42   0.000430  0.928571  0.000860       False     1239      90689   \n",
       "18    29   0.000051  1.000000  0.000102       False        2     569721   \n",
       "19    24   0.000985  0.958333  0.001968       False    80840      23352   \n",
       "20    39   0.001225  1.000000  0.002447       False  3028117      31831   \n",
       "21     4   0.004292  1.000000  0.008547       False  2818505        932   \n",
       "22    45   0.000655  1.000000  0.001309       False    28211      68707   \n",
       "23    10   0.002548  1.000000  0.005084       False  1783257       3924   \n",
       "24    36   0.008910  0.972222  0.017659       False   544448       3928   \n",
       "25    18   0.000061  0.944444  0.000122       False  3379134     278020   \n",
       "26    11   0.000019  1.000000  0.000039       False        2     569721   \n",
       "27    48   0.000082  0.979167  0.000165       False        2     569721   \n",
       "28    14   0.000050  1.000000  0.000101       False  3379134     278020   \n",
       "29     7   0.000025  1.000000  0.000050       False  3379134     278020   \n",
       "30    35   0.000157  1.000000  0.000313       False  1783272     223458   \n",
       "31    27   0.000378  0.962963  0.000757       False    28211      68707   \n",
       "32    20   0.005736  1.000000  0.011406       False   203692       3487   \n",
       "33    40   0.000819  0.800000  0.001636       False    91347      39072   \n",
       "34    32   0.000134  0.937500  0.000268       False  1783272     223458   \n",
       "35    57   0.000624  1.000000  0.001247       False     1760      91371   \n",
       "36    15   0.000026  1.000000  0.000053       False        2     569721   \n",
       "37    27   0.000393  1.000000  0.000786       False    28211      68707   \n",
       "38    19   0.000033  1.000000  0.000067       False        2     569721   \n",
       "39    16   0.000176  1.000000  0.000353       False     1239      90689   \n",
       "40     3   0.000013  1.000000  0.000027       False  1783272     223458   \n",
       "41    78   0.000336  0.961538  0.000671       False  1783272     223458   \n",
       "42    12   0.000047  0.916667  0.000094       False     1224     234365   \n",
       "43     8   0.000029  1.000000  0.000058       False  3379134     278020   \n",
       "44    21   0.000037  1.000000  0.000074       False        2     569721   \n",
       "45    28   0.000125  1.000000  0.000251       False  1783272     223458   \n",
       "46    11   0.000019  1.000000  0.000039       False        2     569721   \n",
       "47    54   0.000180  0.925926  0.000360       False  3379134     278020   \n",
       "\n",
       "                         mrca_name  \n",
       "0                   Pseudomonadati  \n",
       "1              Alphaproteobacteria  \n",
       "2                        Bacillota  \n",
       "3                     Cyanophyceae  \n",
       "4                     Bacteroidota  \n",
       "5                   Actinomycetota  \n",
       "6                   Pseudomonadota  \n",
       "7                         Bacteria  \n",
       "8                         Bacteria  \n",
       "9                   Pseudomonadota  \n",
       "10                      Clostridia  \n",
       "11                  Pseudomonadota  \n",
       "12                  Pseudomonadota  \n",
       "13             Gammaproteobacteria  \n",
       "14                      Bacillales  \n",
       "15                   Actinomycetes  \n",
       "16  Bacteroidota/Chlorobiota group  \n",
       "17                       Bacillota  \n",
       "18                        Bacteria  \n",
       "19                 Burkholderiales  \n",
       "20                    Cyanophyceae  \n",
       "21                     Myxococcota  \n",
       "22             Alphaproteobacteria  \n",
       "23                       PVC group  \n",
       "24                  Mycoplasmatota  \n",
       "25                  Pseudomonadati  \n",
       "26                        Bacteria  \n",
       "27                        Bacteria  \n",
       "28                  Pseudomonadati  \n",
       "29                  Pseudomonadati  \n",
       "30                       Bacillati  \n",
       "31             Alphaproteobacteria  \n",
       "32                    Spirochaetia  \n",
       "33                Enterobacterales  \n",
       "34                       Bacillati  \n",
       "35                   Actinomycetes  \n",
       "36                        Bacteria  \n",
       "37             Alphaproteobacteria  \n",
       "38                        Bacteria  \n",
       "39                       Bacillota  \n",
       "40                       Bacillati  \n",
       "41                       Bacillati  \n",
       "42                  Pseudomonadota  \n",
       "43                  Pseudomonadati  \n",
       "44                        Bacteria  \n",
       "45                       Bacillati  \n",
       "46                        Bacteria  \n",
       "47                  Pseudomonadati  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_stats_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "363473e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['Aliarcobacter butzleri RM4018',\n",
       "  'Helicobacter hepaticus ATCC 51449',\n",
       "  'Thermocrinis albus DSM 14484',\n",
       "  'Hydrogenivirga sp. 128-5-R1-1',\n",
       "  'Sulfurimonas autotrophica DSM 16294',\n",
       "  'Nitratiruptor sp. SB155-2',\n",
       "  'Helicobacter canis NCTC 12740',\n",
       "  'Helicobacter macacae MIT 99-5501',\n",
       "  'Sulfurimonas hongkongensis',\n",
       "  'Helicobacter canadensis MIT 98-5491',\n",
       "  'Helicobacter bilis ATCC 43879',\n",
       "  'Hippea maritima DSM 10411',\n",
       "  'Helicobacter fennelliae MRY12-0050',\n",
       "  'Arcobacter nitrofigilis DSM 7299',\n",
       "  'Desulfurobacterium thermolithotrophum DSM 11699',\n",
       "  'Sulfurimonas denitrificans DSM 1251',\n",
       "  'Sulfurovum sp. AR',\n",
       "  'Nitratifractor salsuginis DSM 16511',\n",
       "  'Campylobacter curvus 525.92',\n",
       "  'Mucispirillum schaedleri ASF457',\n",
       "  'Campylobacter jejuni subsp. jejuni NCTC 11168 = ATCC 700819',\n",
       "  'Aquifex aeolicus VF5',\n",
       "  'Sulfurimonas gotlandica GD1',\n",
       "  'Chitinivibrio alkaliphilus ACht1',\n",
       "  'Hydrogenobaculum sp. HO',\n",
       "  'Borreliella burgdorferi B31',\n",
       "  'Helicobacter bizzozeronii CIII-1',\n",
       "  'Thermodesulfobium narugense DSM 14796',\n",
       "  'Thermodesulfobacterium geofontis OPF15',\n",
       "  'Helicobacter mustelae 12198',\n",
       "  'Sulfurospirillum deleyianum DSM 6946',\n",
       "  'Borrelia duttonii CR2A',\n",
       "  'Helicobacter pylori 26695',\n",
       "  'Sulfurihydrogenibium azorense Az-Fu1',\n",
       "  'Thiovulum sp. ES',\n",
       "  'Lawsonia intracellularis PHE/MN1-00',\n",
       "  'Campylobacter hominis ATCC BAA-381',\n",
       "  'Persephonella marina EX-H1',\n",
       "  'Hydrogenobacter thermophilus TK-6',\n",
       "  'Sulfurihydrogenibium yellowstonense SS-5',\n",
       "  'Campylobacter gracilis RM3268',\n",
       "  'Campylobacter ureolyticus ACS-301-V-Sch3b'],\n",
       " 1: ['Agrobacterium fabrum str. C58',\n",
       "  'Azospirillum sp. B510',\n",
       "  'Tistrella mobilis KA081020-065',\n",
       "  'Mesorhizobium sp. LSHC420B00',\n",
       "  'Sinorhizobium fredii NGR234',\n",
       "  'Nitratireductor pacificus pht-3B',\n",
       "  'Rhizobium etli CFN 42',\n",
       "  'Methylobacterium nodulans ORS 2060',\n",
       "  'Bradyrhizobium diazoefficiens USDA 110',\n",
       "  'Pseudoroseomonas cervicalis ATCC 49957',\n",
       "  'Pelagibacterium halotolerans B2',\n",
       "  'Bradyrhizobium sp. ORS 278',\n",
       "  'Mesorhizobium australicum WSM2073',\n",
       "  'Bradyrhizobium sp. STM 3843',\n",
       "  'Acetobacteraceae bacterium AT-5844',\n",
       "  'Rhodopseudomonas palustris CGA009',\n",
       "  'Pseudorhizobium banfieldiae',\n",
       "  'Sinorhizobium meliloti 1021',\n",
       "  'Mesorhizobium alhagi CCNWXJ12-2',\n",
       "  'Afipia sp. P52-10',\n",
       "  'Azorhizobium caulinodans ORS 571',\n",
       "  'Mesorhizobium amorphae CCNWGS0123',\n",
       "  'Rhizobium mesoamericanum STM3625',\n",
       "  'Nitratireductor indicus C115',\n",
       "  'Brucella anthropi ATCC 49188',\n",
       "  'Afipia broomeae ATCC 49717',\n",
       "  'Rhizobium sp. PDO1-076',\n",
       "  'Brucella intermedia 229E',\n",
       "  'Phyllobacterium sp. YR531',\n",
       "  'Xanthobacter versatilis',\n",
       "  'Allorhizobium ampelinum S4'],\n",
       " 2: ['Lachnospiraceae bacterium A4',\n",
       "  'Subdoligranulum variabile DSM 15176',\n",
       "  'Anaerotruncus sp. G3(2012)',\n",
       "  'Lachnospiraceae bacterium 10-1',\n",
       "  'Lachnospiraceae bacterium A2',\n",
       "  'Eubacterium sp. 14-2',\n",
       "  'Lachnospiraceae bacterium 3-1',\n",
       "  'Mediterraneibacter gnavus CC55_001C',\n",
       "  'Lachnospiraceae bacterium 9_1_43BFAA',\n",
       "  'Pseudoflavonifractor capillosus ATCC 29799',\n",
       "  'Clostridium sp. D5',\n",
       "  'Oribacterium sinus F0268',\n",
       "  'Lachnospiraceae bacterium 1_4_56FAA',\n",
       "  'Firmicutes bacterium ASF500',\n",
       "  'Catonella morbi ATCC 51271',\n",
       "  'Lachnospiraceae bacterium 2_1_46FAA',\n",
       "  'Butyrivibrio proteoclasticus B316',\n",
       "  'butyrate-producing bacterium SM4/1',\n",
       "  'Faecalibacterium prausnitzii L2-6',\n",
       "  'Ruminococcaceae bacterium D16',\n",
       "  '[Clostridium] symbiosum WAL-14163',\n",
       "  'Clostridium sp. MD294',\n",
       "  'Eubacterium plexicaudatum ASF492',\n",
       "  'Roseburia hominis A2-183',\n",
       "  'Lachnospira eligens ATCC 27750',\n",
       "  'Lachnospiraceae bacterium MD308',\n",
       "  'Lachnospiraceae oral taxon 107 str. F0167',\n",
       "  'Lachnospiraceae bacterium 3_1_57FAA_CT1',\n",
       "  '[Clostridium] methylpentosum DSM 5476',\n",
       "  'Butyricicoccus pullicaecorum 1.2',\n",
       "  'Lachnospiraceae bacterium MD335',\n",
       "  'Lachnospiraceae bacterium M18-1',\n",
       "  'Oscillibacter sp. 1-3',\n",
       "  'Lachnospiraceae bacterium 28-4',\n",
       "  '[Ruminococcus] lactaris ATCC 29176',\n",
       "  'Lachnospiraceae bacterium 5_1_57FAA',\n",
       "  'Oribacterium parvum ACB1',\n",
       "  'Hungatella hathewayi WAL-18680',\n",
       "  '[Eubacterium] cellulosolvens 6',\n",
       "  'Marvinbryantia formatexigens DSM 14469'],\n",
       " 3: ['Synechococcus sp. RCC307',\n",
       "  'Synechococcus sp. CC9311',\n",
       "  'Prochlorococcus marinus str. MIT 9211',\n",
       "  'Synechococcus sp. PCC 6312',\n",
       "  'Prochlorococcus marinus subsp. marinus str. CCMP1375',\n",
       "  'Prochlorococcus marinus str. MIT 9313',\n",
       "  'Synechococcus sp. WH 7803',\n",
       "  'Rubidibacter lacunae KORDI 51-2',\n",
       "  'Cyanobacterium aponinum PCC 10605',\n",
       "  'Synechococcus sp. PCC 7502',\n",
       "  'Cyanobacterium stanieri PCC 7202',\n",
       "  'Thermosynechococcus vestitus BP-1',\n",
       "  \"'Nostoc azollae' 0708\",\n",
       "  'Prochlorococcus marinus str. MIT 9301',\n",
       "  'Prochlorococcus marinus str. NATL2A',\n",
       "  'Candidatus Atelocyanobacterium thalassa isolate ALOHA',\n",
       "  'Synechococcus sp. RS9916',\n",
       "  'Synechococcus sp. WH 5701',\n",
       "  'Picosynechococcus sp. PCC 7002',\n",
       "  'Cyanobium sp. PCC 7001',\n",
       "  \"Synechococcus sp. JA-2-3B'a(2-13)\",\n",
       "  'Synechocystis sp. PCC 6803 substr. Kazusa'],\n",
       " 4: ['Segatella baroniae F0067',\n",
       "  'Odoribacter splanchnicus DSM 20712',\n",
       "  'Porphyromonas endodontalis ATCC 35406',\n",
       "  'Hoylesella marshii DSM 16973 = JCM 13450',\n",
       "  'Porphyromonas asaccharolytica DSM 20707',\n",
       "  'Candidatus Azobacteroides pseudotrichonymphae genomovar. CFP2',\n",
       "  'Segatella baroniae B14',\n",
       "  'Prevotella bivia DSM 20514',\n",
       "  'Hallella multisaccharivorax DSM 17128',\n",
       "  'Bacteroides coprosuis DSM 18011',\n",
       "  'Prevotella histicola F0411',\n",
       "  'Bacteroides helcogenes P 36-108',\n",
       "  'Bacteroidetes oral taxon 274 str. F0058',\n",
       "  'Paludibacter propionicigenes WB4',\n",
       "  'Porphyromonas gingivalis W83',\n",
       "  'Phocaeicola massiliensis B84634 = Timone 84634 = DSM 17679 = JCM 13223',\n",
       "  'Alistipes shahii WAL 8301',\n",
       "  'Parabacteroides distasonis ATCC 8503',\n",
       "  'Prevotella dentalis DSM 3688',\n",
       "  'Dysgonomonas gadei ATCC BAA-286',\n",
       "  'Dysgonomonas mossii DSM 22836',\n",
       "  'Parabacteroides johnsonii CL02T12C29',\n",
       "  'Segatella oris C735',\n",
       "  'Alloprevotella tannerae ATCC 51259',\n",
       "  'Prevotella sp. oral taxon 317 str. F0108',\n",
       "  'Bacteroides thetaiotaomicron VPI-5482',\n",
       "  'Segatella salivae DSM 15606',\n",
       "  'Bacteroides fluxus YIT 12057',\n",
       "  'Bacteroidetes bacterium oral taxon 272 str. F0290',\n",
       "  'Bacteroides fragilis CL05T12C13',\n",
       "  'Prevotella veroralis F0319',\n",
       "  'Hoylesella pleuritidis F0068',\n",
       "  'Prevotella sp. oral taxon 299 str. F0039',\n",
       "  'Segatella maculosa OT 289',\n",
       "  'Hallella bergensis DSM 17361',\n",
       "  'Alloprevotella rava F0323',\n",
       "  'Prevotella micans F0438',\n",
       "  'Bacteroides sp. D2',\n",
       "  'Porphyromonas sp. oral taxon 279 str. F0450',\n",
       "  'Segatella buccae ATCC 33574',\n",
       "  'Xylanibacter ruminicola 23',\n",
       "  'Bacteroides nordii CL02T12C05',\n",
       "  'Hoylesella timonensis CRIS 5C-B1',\n",
       "  'Hoylesella oralis ATCC 33269',\n",
       "  'Alloprevotella sp. oral taxon 473 str. F0040',\n",
       "  'Segatella oulorum F0390',\n",
       "  'Alistipes indistinctus YIT 12060',\n",
       "  'Tannerella forsythia 92A2',\n",
       "  'Bacteroides reticulotermitis JCM 10512',\n",
       "  'Bacteroides uniformis dnLKV2',\n",
       "  'Hoylesella saccharolytica F0055',\n",
       "  'Prevotella multiformis DSM 16608'],\n",
       " 5: ['Pseudonocardia dioxanivorans CB1190',\n",
       "  'Prescottella equi ATCC 33707',\n",
       "  'Mycobacterium tuberculosis H37Rv',\n",
       "  'Mycobacterium tuberculosis variant bovis AF2122/97',\n",
       "  'Patulibacter medicamentivorans',\n",
       "  'Aeromicrobium marinum DSM 15272',\n",
       "  'Gordonia amarae NBRC 15530',\n",
       "  'Nocardia brasiliensis ATCC 700358',\n",
       "  'Gordonia aichiensis NBRC 108223',\n",
       "  'Mycobacterium avium subsp. paratuberculosis K-10',\n",
       "  'Mycolicibacter sinensis',\n",
       "  'Segniliparus rugosus ATCC BAA-974',\n",
       "  'Mycolicibacterium chubuense NBB4',\n",
       "  'Mycolicibacterium hassiacum DSM 44199',\n",
       "  'Rhodococcus rhodnii LMG 5362',\n",
       "  'Mycobacterium tuberculosis H37Ra',\n",
       "  'Mycobacterium tuberculosis CDC1551',\n",
       "  'Nocardia farcinica IFM 10152',\n",
       "  'Mycolicibacterium thermoresistibile ATCC 19527',\n",
       "  'Nocardia cyriacigeorgica GUH-2',\n",
       "  'Saccharomonospora marina XMU15',\n",
       "  'Mycolicibacterium neoaurum VKM Ac-1815D',\n",
       "  'Gordonia polyisoprenivorans VH2',\n",
       "  'Nocardia asteroides NBRC 15531',\n",
       "  'Gordonia soli NBRC 108243',\n",
       "  'Segniliparus rotundus DSM 44985',\n",
       "  'Mycolicibacterium rhodesiae NBB3',\n",
       "  'Gordonia bronchialis DSM 43247',\n",
       "  'Hoyosella subflava DQS3-9A1',\n",
       "  'Gordonia rhizosphera NBRC 16068',\n",
       "  'Nocardia nova SH22a',\n",
       "  'Mycolicibacterium smegmatis MC2 155',\n",
       "  'Mycobacterium marinum M',\n",
       "  'Tsukamurella paurometabola DSM 20162',\n",
       "  'Mycobacterium leprae TN'],\n",
       " 6: ['Alcanivorax borkumensis SK2',\n",
       "  'Acinetobacter rudis CIP 110305',\n",
       "  'Candidatus Paraburkholderia kirkii UZHbot1',\n",
       "  'Kangiella koreensis DSM 16069',\n",
       "  'Acinetobacter pittii PHEA-2',\n",
       "  'Psychrobacter arcticus 273-4',\n",
       "  'Pelistega indica',\n",
       "  'gamma proteobacterium HTCC5015',\n",
       "  'Acinetobacter brisouii CIP 110357',\n",
       "  'Polynucleobacter asymbioticus QLW-P1DMWA-1',\n",
       "  'Acinetobacter guillouiae NIPH 991',\n",
       "  'Xylella fastidiosa Temecula1',\n",
       "  'Acinetobacter indicus CIP 110367'],\n",
       " 7: ['Nocardioides sp. JS614',\n",
       "  'marine actinobacterium PHSC20C1',\n",
       "  'Conexibacter woesei DSM 14684',\n",
       "  'Kineococcus radiotolerans SRS30216 = ATCC BAA-149',\n",
       "  'Acidothermus cellulolyticus 11B',\n",
       "  'Arthrobacter nitrophenolicus',\n",
       "  'Brachybacterium faecium DSM 4810',\n",
       "  'Saccharomonospora viridis DSM 43017',\n",
       "  'Saccharomonospora glauca K62',\n",
       "  'Janibacter sp. HTCC2649',\n",
       "  'Arthrobacter globiformis NBRC 12137',\n",
       "  'Beutenbergia cavernae DSM 12333',\n",
       "  'Renibacterium salmoninarum ATCC 33209',\n",
       "  'Cellulomonas flavigena DSM 20109',\n",
       "  'Nocardioides sp. CF8',\n",
       "  'Candidatus Protofrankia datiscae',\n",
       "  'Arthrobacter crystallopoietes BAB-32',\n",
       "  'Intrasporangium calvum DSM 43043',\n",
       "  'Microbacterium maritypicum MF109',\n",
       "  'Phycicoccus elongatus Lp2',\n",
       "  'Agrococcus pavilionensis RW1',\n",
       "  'Microlunatus phosphovorus NM-1',\n",
       "  'Brevibacterium casei S18',\n",
       "  'Austwickia chelonae NBRC 105200',\n",
       "  'Blastococcus saxobsidens DD2',\n",
       "  'Deinococcus deserti VCD115',\n",
       "  'Nakamurella multipartita DSM 44233',\n",
       "  'Kineosphaera limosa NBRC 100340',\n",
       "  'Cellulomonas gilvus ATCC 13127',\n",
       "  'Modestobacter italicus',\n",
       "  'Isoptericola variabilis 225',\n",
       "  'Geodermatophilus obscurus DSM 43160',\n",
       "  'Mobilicoccus pelagius NBRC 104925',\n",
       "  'Cellulomonas fimi ATCC 484',\n",
       "  'Arthrobacter sp. FB24',\n",
       "  'Xylanimonas cellulosilytica DSM 15894',\n",
       "  'Ilumatobacter coccineus YM16-304',\n",
       "  'Acidimicrobium ferrooxidans DSM 10331',\n",
       "  'Thermobispora bispora DSM 43833'],\n",
       " 8: ['Halanaerobium saccharolyticum subsp. saccharolyticum DSM 6643',\n",
       "  'Dethiosulfovibrio peptidovorans DSM 11002',\n",
       "  'Pseudothermotoga thermarum DSM 5069',\n",
       "  'Halanaerobium praevalens DSM 2228',\n",
       "  'Aminomonas paucivorans DSM 12260',\n",
       "  'Ilyobacter polytropus DSM 2926',\n",
       "  'Acetomicrobium mobile DSM 13181',\n",
       "  'Thermanaerovibrio acidaminovorans DSM 6589',\n",
       "  'Thermovirga lienii DSM 17291',\n",
       "  'Fusobacterium animalis ATCC 51191',\n",
       "  'Aminobacterium colombiense DSM 12261',\n",
       "  'Jonquetella anthropi DSM 22815',\n",
       "  'Halanaerobium hydrogeniformans'],\n",
       " 9: ['Bermanella marisrubri',\n",
       "  'Alloalcanivorax dieselolei B5',\n",
       "  'Cycloclasticus sp. P1',\n",
       "  'Betaproteobacteria bacterium MOLA814',\n",
       "  'gamma proteobacterium HdN1',\n",
       "  'Laribacter hongkongensis HLHK9',\n",
       "  'Lautropia mirabilis ATCC 51599',\n",
       "  'Halomonas sp. BC04',\n",
       "  'Salinisphaera shabanensis E1L3A',\n",
       "  'Isoalcanivorax pacificus W11-5',\n",
       "  'Neptuniibacter caesariensis',\n",
       "  'Limnobacter sp. MED105',\n",
       "  'Litchfieldella anticariensis FP35 = DSM 16096',\n",
       "  'Spiribacter curvatus',\n",
       "  'Halomonas sp. GFAJ-1',\n",
       "  'Thalassolituus oleivorans MIL-1',\n",
       "  'Marinobacter sp. BSs20148',\n",
       "  'Halomonas huangheensis',\n",
       "  'Chromohalobacter israelensis DSM 3043',\n",
       "  'gamma proteobacterium IMCC1989'],\n",
       " 10: ['Clostridium cellulovorans 743B',\n",
       "  'Acetivibrio clariflavus DSM 19732',\n",
       "  'Ruminococcus albus 8',\n",
       "  'Acetivibrio thermocellus ATCC 27405',\n",
       "  'Mahella australiensis 50-1 BON',\n",
       "  'Acetivibrio straminisolvens JCM 21531',\n",
       "  'Ruminiclostridium cellulolyticum H10',\n",
       "  'Cellulosilyticum lentocellum DSM 5427',\n",
       "  'Ruminococcus champanellensis 18P13 = JCM 17042',\n",
       "  'Thermoclostridium stercorarium subsp. stercorarium DSM 8532',\n",
       "  'Ruminiclostridium cellobioparum subsp. termitidis CT1112',\n",
       "  'Lachnoclostridium phytofermentans ISDg',\n",
       "  'Ruminococcus flavefaciens 007c'],\n",
       " 11: ['Methylobacter tundripaludum SV96',\n",
       "  'Marichromatium purpuratum 984',\n",
       "  'Methylobacillus flagellatus KT',\n",
       "  'Methylotuvimicrobium alcaliphilum 20Z',\n",
       "  'endosymbiont of Tevnia jerichonana (vent Tica)',\n",
       "  'Methylophaga frappieri',\n",
       "  'Nitrococcus mobilis Nb-231',\n",
       "  'Nitrosomonas europaea ATCC 19718',\n",
       "  'Halothiobacillus neapolitanus c2',\n",
       "  'Methylophaga thiooxydans DMS010',\n",
       "  'Methylotenera mobilis JLW8',\n",
       "  'Methylomicrobium album BG8',\n",
       "  'Thioalkalivibrio paradoxus ARh 1',\n",
       "  'Nitrosospira multiformis ATCC 25196',\n",
       "  'Sulfuricella denitrificans skB26',\n",
       "  'endosymbiont of unidentified scaly snail isolate Monju',\n",
       "  'Thiocystis violascens DSM 198',\n",
       "  'Allochromatium vinosum DSM 180',\n",
       "  'Methyloglobulus morosus KoM1',\n",
       "  'Acidithiobacillus caldus SM-1',\n",
       "  'Magnetococcus marinus MC-1',\n",
       "  'Methylococcus capsulatus str. Bath',\n",
       "  'Nitrosococcus oceani ATCC 19707',\n",
       "  'Methylomonas methanica MC09',\n",
       "  'Thioflavicoccus mobilis 8321',\n",
       "  'Alkalilimnicola ehrlichii MLHE-1',\n",
       "  'Thiobacillus denitrificans ATCC 25259',\n",
       "  'Thiorhodovibrio frisius',\n",
       "  'Gallionella capsiferriformans ES-2',\n",
       "  'gamma proteobacterium IMCC2047',\n",
       "  'Thiocapsa marina 5811',\n",
       "  'Acidithiobacillus ferrooxidans ATCC 23270',\n",
       "  'Thioalkalivibrio sulfidiphilus HL-EbGr7',\n",
       "  'Methylophaga aminisulfidivorans MP',\n",
       "  'Beggiatoa alba B18LD',\n",
       "  'Nitrosomonas sp. AL212',\n",
       "  'Nitrosococcus halophilus Nc 4',\n",
       "  'Methylotenera versatilis 301',\n",
       "  'Candidatus Symbiobacter mobilis CR',\n",
       "  'Methylophaga lonarensis MPL',\n",
       "  'Mariprofundus ferrooxydans PV-1',\n",
       "  'Methylovorus glucosotrophus SIP3-4',\n",
       "  'Sideroxydans lithotrophicus ES-1',\n",
       "  'Thiothrix nivea DSM 5205',\n",
       "  'Halorhodospira halophila SL1',\n",
       "  'Nitrosospira lacus'],\n",
       " 12: ['Caulobacter vibrioides NA1000',\n",
       "  'Hirschia baltica ATCC 49814',\n",
       "  'alpha proteobacterium IMCC14465',\n",
       "  'Gammaproteobacteria bacterium MOLA455',\n",
       "  'Xanthomonas campestris pv. campestris str. ATCC 33913',\n",
       "  'gamma proteobacterium BDW918',\n",
       "  'Pseudoxanthomonas spadix BD-a59',\n",
       "  'Stenotrophomonas maltophilia K279a',\n",
       "  'Oceanicaulis sp. HTCC2633',\n",
       "  'Luminiphilus syltensis NOR5-1B',\n",
       "  'Asticcacaulis sp. YBE204',\n",
       "  'Xanthomonas albilineans GPE PC73',\n",
       "  'Brevundimonas subvibrioides ATCC 15264',\n",
       "  'Asticcacaulis sp. AC402',\n",
       "  'marine gamma proteobacterium HTCC2080',\n",
       "  'gamma proteobacterium HIMB55',\n",
       "  'Frateuria aurantia DSM 6220',\n",
       "  'Xanthomonas oryzae pv. oryzae KACC 10331',\n",
       "  'Aequoribacter fuscus',\n",
       "  'Congregibacter litoralis KT71',\n",
       "  'Hydrocarboniphaga effusa AP103',\n",
       "  'Maricaulis maris MCS10',\n",
       "  'marine gamma proteobacterium HTCC2143',\n",
       "  'Hyphomonas neptunium ATCC 15444',\n",
       "  'Asticcacaulis benevestitus DSM 16100 = ATCC BAA-896',\n",
       "  'marine gamma proteobacterium HTCC2148',\n",
       "  'Massilia timonae CCUG 45783',\n",
       "  'Pseudoxanthomonas suwonensis 11-1',\n",
       "  'SAR86 cluster bacterium SAR86E',\n",
       "  'Rhodanobacter fulvus Jip2',\n",
       "  'Caulobacter vibrioides CB15',\n",
       "  'Asticcacaulis sp. AC466',\n",
       "  'Acinetobacter gerneri DSM 14967 = CIP 107464 = MTCC 9824',\n",
       "  'Rhodanobacter spathiphylli B39'],\n",
       " 13: ['Vibrio azureus NBRC 104587',\n",
       "  'Shewanella woodyi ATCC 51908',\n",
       "  'Ferrimonas balearica DSM 9799',\n",
       "  'Catenovulum agarivorans DS-2',\n",
       "  'Shewanella amazonensis SB2B',\n",
       "  'Idiomarina loihiensis L2TR',\n",
       "  'Idiomarina xiamenensis 10-D-4',\n",
       "  'Aliivibrio fischeri ES114',\n",
       "  'Oceanimonas sp. GK1',\n",
       "  'Gallaecimonas xiamenensis 3-C-1',\n",
       "  'Shewanella loihica PV-4',\n",
       "  'Teredinibacter turnerae T7901',\n",
       "  'Vibrio cholerae O1 biovar El Tor str. N16961',\n",
       "  'Tolumonas auensis DSM 9187',\n",
       "  'Aliiglaciecola lipolytica E3',\n",
       "  'Plesiomonas shigelloides 302-73',\n",
       "  'Paraglaciecola arctica BSs20135',\n",
       "  'Vibrio ezurae NBRC 102218',\n",
       "  'Cellvibrio japonicus Ueda107',\n",
       "  'Shewanella sediminis HAW-EB3',\n",
       "  'Photobacterium profundum SS9',\n",
       "  'Reinekea blandensis MED297',\n",
       "  'Paraglaciecola psychrophila 170',\n",
       "  'Vibrio nigripulchritudo',\n",
       "  'Shewanella oneidensis MR-1',\n",
       "  'Rheinheimera nanhaiensis E407-8',\n",
       "  'Grimontia hollisae CIP 101886',\n",
       "  'Brumicola nitratireducens FR1064',\n",
       "  'Psychromonas ingrahamii 37',\n",
       "  'Vibrio caribbeanicus ATCC BAA-2122',\n",
       "  'Simiduia agarivorans SA1 = DSM 21679',\n",
       "  'Shewanella pealeana ATCC 700345',\n",
       "  'Shewanella frigidimarina NCIMB 400',\n",
       "  'Photobacterium damselae subsp. damselae CIP 102761',\n",
       "  'Shewanella violacea DSS12',\n",
       "  'Pseudoalteromonas translucida TAC125',\n",
       "  'Psychromonas sp. CNPT3',\n",
       "  'Shewanella denitrificans OS217',\n",
       "  'Aeromonas hydrophila subsp. hydrophila ATCC 7966',\n",
       "  'Saccharophagus degradans 2-40'],\n",
       " 14: ['Brevibacillus brevis NBRC 100599',\n",
       "  'Paenibacillus dendritiformis C454',\n",
       "  'Paenibacillus sp. JCM 10914',\n",
       "  'Brevibacillus laterosporus LMG 15441',\n",
       "  'Brevibacillus borstelensis AK1',\n",
       "  'Paenibacillus sp. HGF7',\n",
       "  'Paenibacillus mucilaginosus 3016',\n",
       "  'Halalkalibacter hemicellulosilyticusJCM 9152',\n",
       "  'Thermobacillus composti KWC4',\n",
       "  'Brevibacillus panacihumi W25',\n",
       "  'Paenibacillus sp. oral taxon 786 str. D14',\n",
       "  'Paenibacillus pini JCM 16418'],\n",
       " 15: ['Streptomyces coelicolor A3(2)',\n",
       "  'Streptomyces turgidiscabies Car8',\n",
       "  'Streptomyces bingchenggensis BCW-1',\n",
       "  'Streptomyces sp. SirexAA-E',\n",
       "  'Kitasatospora setae KM-6054',\n",
       "  'Actinoplanes missouriensis 431',\n",
       "  'Streptomyces zinciresistens K42',\n",
       "  'Streptomyces afghaniensis 772',\n",
       "  'Microbispora sp. ATCC PTA-5024',\n",
       "  'Streptosporangium roseum DSM 43021',\n",
       "  'Saccharothrix espanaensis DSM 44229',\n",
       "  'Thermomonospora curvata DSM 43183',\n",
       "  'Actinoplanes sp. N902-109',\n",
       "  'Streptomyces auratus AGR0001',\n",
       "  'Pseudofrankia inefficax',\n",
       "  'Amycolatopsis keratiniphila',\n",
       "  'Saccharopolyspora erythraea NRRL 2338',\n",
       "  'Streptomyces ipomoeae 91-03',\n",
       "  'Actinoplanes friuliensis DSM 7358',\n",
       "  'Streptomyces collinus Tu 365',\n",
       "  'Salinispora tropica CNB-440',\n",
       "  'Streptomyces venezuelae ATCC 10712',\n",
       "  'Streptomyces viridochromogenes DSM 40736',\n",
       "  'Stackebrandtia nassauensis DSM 44728',\n",
       "  'Streptantibioticus cattleyicolor NRRL 8057 = DSM 46488',\n",
       "  'Amycolatopsis mediterranei U32',\n",
       "  'Kutzneria albida DSM 43870',\n",
       "  'Streptomycetaceae bacterium MP113-05',\n",
       "  'Streptomyces roseochromogenus subsp. oscitans DS 12.976',\n",
       "  'Nocardiopsis dassonvillei subsp. dassonvillei DSM 43111',\n",
       "  'Streptomyces sp. HPH0547',\n",
       "  'Streptomyces avermitilis MA-4680 = NBRC 14893',\n",
       "  'Actinoplanes sp. SE50/110',\n",
       "  'Streptomyces davaonensis JCM 4913',\n",
       "  'Kribbella flavida DSM 17836',\n",
       "  'Actinosynnema mirum DSM 43827',\n",
       "  'Frankia casuarinae',\n",
       "  'Streptomyces clavuligerus',\n",
       "  'Micromonospora sp. ATCC 39149',\n",
       "  'Streptomyces scabiei 87.22',\n",
       "  'Streptomyces pristinaespiralis ATCC 25486',\n",
       "  'Catenulispora acidiphila DSM 44928',\n",
       "  'Frankia alni ACN14a',\n",
       "  'Streptomyces griseoaurantiacus M045',\n",
       "  'Streptomyces aurantiacus JA 4570',\n",
       "  'Micromonospora lupini str. Lupac 08',\n",
       "  'Actinokineospora spheciospongiae',\n",
       "  'Streptomyces sp. HGB0020'],\n",
       " 16: ['Polaribacter irgensii 23-P',\n",
       "  'Riemerella anatipestifer RA-CH-1',\n",
       "  'Sphingobacterium spiritivorum ATCC 33861',\n",
       "  'Weeksella virosa DSM 16922',\n",
       "  'Echinicola vietnamensis DSM 17526',\n",
       "  'Galbibacter orientalis DSM 19592',\n",
       "  'Marivirga tractuosa DSM 4126',\n",
       "  'Bergeyella zoohelcum ATCC 43767',\n",
       "  'Indibacter alkaliphilus LW1',\n",
       "  'Imtechella halotolerans K1',\n",
       "  'Algoriphagus machipongonensis',\n",
       "  'Bernardetia litoralis DSM 6794',\n",
       "  'Belliella baltica DSM 15883',\n",
       "  'Gillisia limnaea DSM 15749',\n",
       "  'Capnocytophaga canimorsus Cc5',\n",
       "  'Dyadobacter fermentans DSM 18053',\n",
       "  'Zobellia galactanivorans',\n",
       "  'Zhouia amylolytica AD3',\n",
       "  'Allomuricauda ruestringensis DSM 13258',\n",
       "  'Capnocytophaga sp. oral taxon 332 str. F0381',\n",
       "  'Flavobacteriaceae bacterium 3519-10',\n",
       "  'Lacinutrix sp. 5H-3-7-4',\n",
       "  'Rhodonellum psychrophilum GCM71 = DSM 17998',\n",
       "  'Sphingobacterium paucimobilis HER1398',\n",
       "  'Polaribacter sp. MED152',\n",
       "  'Capnocytophaga ochracea DSM 7271',\n",
       "  'Cytophaga hutchinsonii ATCC 33406',\n",
       "  'Cesiribacter andamanensis AMV16',\n",
       "  'Cyclobacterium marinum DSM 745',\n",
       "  'Pedobacter heparinus DSM 2366',\n",
       "  'Flavobacterium psychrophilum JIP02/86',\n",
       "  'Psychroflexus gondwanensis ACAM 44',\n",
       "  'Haliscomenobacter hydrossis DSM 1100',\n",
       "  'Cecembia lonarensis LW9',\n",
       "  'Maribacter sp. HTCC2170',\n",
       "  'Fulvivirga imtechensis AK7',\n",
       "  'Leeuwenhoekiella blandensis MED217',\n",
       "  'Fibrella aestuarina BUZ 2',\n",
       "  'Microscilla marina ATCC 23134',\n",
       "  'Capnocytophaga sp. oral taxon 338 str. F0234',\n",
       "  'Formosa agariphila KMM 3901',\n",
       "  'Mariniradius saccharolyticus AK6',\n",
       "  'Mucilaginibacter paludis DSM 18603',\n",
       "  'Pedobacter sp. BAL39',\n",
       "  'Candidatus Amoebophilus asiaticus 5a2',\n",
       "  'Flavobacteria bacterium BBFL7',\n",
       "  'Ornithobacterium rhinotracheale DSM 15997',\n",
       "  'Flavobacterium indicum GPTSA100-9 = DSM 17447',\n",
       "  'Flavobacteria bacterium MS024-3C',\n",
       "  'Flavobacteriales bacterium ALC-1',\n",
       "  'Solitalea canadensis DSM 3403',\n",
       "  'Flavobacterium columnare ATCC 49512',\n",
       "  'Runella slithyformis DSM 19594',\n",
       "  'Robiginitalea biformata HTCC2501',\n",
       "  'Owenweeksia hongkongensis DSM 17368',\n",
       "  'Flavobacterium branchiophilum FL-15',\n",
       "  'Saprospira grandis str. Lewin',\n",
       "  'Niabella soli DSM 19437',\n",
       "  'Nitritalea halalkaliphila LW7',\n",
       "  'Ignavibacterium album JCM 16511',\n",
       "  'Leadbetterella byssophila DSM 17132',\n",
       "  'Fluviicola taffensis DSM 16823',\n",
       "  'Arcticibacter svalbardensis MN12-7',\n",
       "  'Croceibacter atlanticus HTCC2559',\n",
       "  'Cellulophaga lytica DSM 7489',\n",
       "  'Galbibacter marinus',\n",
       "  'Fibrisoma limi BUZ 3',\n",
       "  'Winogradskyella psychrotolerans RS-3',\n",
       "  'Flavobacteria bacterium BAL38',\n",
       "  'Bizionia argentinensis JUB59',\n",
       "  'Lunatimonas lonarensis',\n",
       "  'Pseudopedobacter saltans DSM 12145'],\n",
       " 17: ['Geobacillus kaustophilus HTA426',\n",
       "  'Alicyclobacillus acidocaldarius subsp. acidocaldarius DSM 446',\n",
       "  'Shouchella clausii KSM-K16',\n",
       "  'Sporolactobacillus laevolacticus DSM 442',\n",
       "  'Bacillus licheniformis DSM 13 = ATCC 14580',\n",
       "  'Bacillus infantis NRRL B-14911',\n",
       "  'Alkalihalobacillus alcalophilus ATCC 27647 = CGMCC 1.3604',\n",
       "  'Neobacillus bataviensis LMG 21833',\n",
       "  'Schinkia azotoformans LMG 9581',\n",
       "  'Bacillus anthracis',\n",
       "  'Amphibacillus xylanus NBRC 15112',\n",
       "  'Oceanobacillus iheyensis HTE831',\n",
       "  'Staphylococcus epidermidis RP62A',\n",
       "  'Thermaerobacter marianensis DSM 12885',\n",
       "  'Bacillus sp. 1NLA3E',\n",
       "  'Fictibacillus macauensis ZFHKF-1',\n",
       "  'Bacillus subtilis subsp. subtilis str. 168',\n",
       "  'Exiguobacterium chiriqhucha RW-2',\n",
       "  'Salimicrobium jeotgali',\n",
       "  'Halalkalibacter wakoensis JCM 9140',\n",
       "  'Priestia megaterium QM B1551',\n",
       "  'Staphylococcus aureus subsp. aureus NCTC 8325',\n",
       "  'Mesobacillus boroniphilus JCM 21738',\n",
       "  'Alkalihalophilus pseudofirmus OF4',\n",
       "  'Halobacillus sp. BAB-2008',\n",
       "  'Viridibacillus arenosi FSL R5-213',\n",
       "  'Gracilibacillus halophilus YIM-C55.5',\n",
       "  'Bacillus smithii 7_3_47FAA',\n",
       "  'Halobacillus halophilus DSM 2266',\n",
       "  'Staphylococcus massiliensis S46',\n",
       "  'Paenisporosarcina sp. HGH0030',\n",
       "  'Halalkalibacterium halodurans C-125',\n",
       "  'Gracilibacillus boraciitolerans JCM 21714',\n",
       "  'Bacillus cereus ATCC 14579',\n",
       "  'Paenibacillus curdlanolyticus YK9',\n",
       "  'Macrococcoides caseolyticum JCSC5402',\n",
       "  'Halalkalibacter akibai JCM 9157',\n",
       "  'Exiguobacterium artemiae 255-15',\n",
       "  'Evansella cellulosilytica DSM 2522',\n",
       "  'Staphylococcus saprophyticus subsp. saprophyticus ATCC 15305 = NCTC 7292',\n",
       "  'Bhargavaea cecembensis DSE10',\n",
       "  'Solibacillus silvestris StLB046'],\n",
       " 18: ['Candidatus Pelagibacter sp. IMCC9063',\n",
       "  'Legionella longbeachae NSW150',\n",
       "  'Bartonella bovis 91-4',\n",
       "  'Bacteriovorax sp. DB6_IX',\n",
       "  'Candidatus Endolissoclinum faulkneri L5',\n",
       "  'Bacteriovorax sp. BSW11_IV',\n",
       "  'Candidatus Babela massiliensis',\n",
       "  'Legionella oakridgensis ATCC 33761 = DSM 21215',\n",
       "  'Bacteriovorax sp. Seq25_V',\n",
       "  'Halobacteriovorax marinus SJ',\n",
       "  'Candidatus Protochlamydia amoebophila UWE25',\n",
       "  'Waddlia chondrophila WSU 86-1044',\n",
       "  'Liberibacter crescens BT-1',\n",
       "  'Coxiella burnetii RSA 493',\n",
       "  'Spiribacter salinus M19-40',\n",
       "  'Candidatus Liberibacter americanus str. Sao Paulo',\n",
       "  'Orientia tsutsugamushi str. Boryong',\n",
       "  'Legionella pneumophila subsp. pneumophila str. Philadelphia 1',\n",
       "  'Legionella drancourtii LLAP12',\n",
       "  'Pseudobdellovibrio exovorus JSS',\n",
       "  'Candidatus Midichloria mitochondrii IricVA',\n",
       "  'Leptospira borgpetersenii str. 200701203',\n",
       "  'Bdellovibrio bacteriovorus HD100',\n",
       "  'Holospora obtusa F1',\n",
       "  'Candidatus Pelagibacter ubique HTCC1062',\n",
       "  'Simkania negevensis Z',\n",
       "  'Micavibrio aeruginosavorus ARL-13',\n",
       "  'Parachlamydia acanthamoebae UV-7',\n",
       "  'Chlamydia trachomatis D/UW-3/CX'],\n",
       " 19: ['Burkholderia multivorans ATCC 17616',\n",
       "  'Acidovorax sp. CF316',\n",
       "  'Paracidovorax avenae ATCC 19860',\n",
       "  'Verminephrobacter eiseniae EF01-2',\n",
       "  'unknown',\n",
       "  'Herbaspirillum seropedicae SmR1',\n",
       "  'Alicycliphilus denitrificans K601',\n",
       "  'Cupriavidus necator H16',\n",
       "  'Cupriavidus metallidurans CH34',\n",
       "  'Delftia acidovorans SPH-1',\n",
       "  'Bordetella pertussis Tohama I',\n",
       "  'Burkholderia pseudomallei K96243',\n",
       "  'Advenella mimigardefordensis DPN7',\n",
       "  'Acidovorax delafieldii 2AN',\n",
       "  'Hylemonella gracilis ATCC 19624',\n",
       "  'Polaromonas sp. JS666',\n",
       "  'Ramlibacter tataouinensis TTB310',\n",
       "  'Paraburkholderia graminis C4D1M',\n",
       "  'Ralstonia sp. 5_2_56FAA',\n",
       "  'Paraburkholderia phymatum STM815',\n",
       "  'Collimonas fungivorans Ter331',\n",
       "  'Alcaligenes sp. EGD-AK7',\n",
       "  'Herbaspirillum sp. YR522',\n",
       "  'Bordetella avium 197N'],\n",
       " 20: ['Phormidium nigroviride PCC 7112',\n",
       "  'Calothrix sp. PCC 6303',\n",
       "  'Chamaesiphon minutus PCC 6605',\n",
       "  'Oxynema acuminatum PCC 6304',\n",
       "  'Nostoc sp. PCC 7524',\n",
       "  'Gloeocapsa sp. PCC 7428',\n",
       "  'Rivularia sp. PCC 7116',\n",
       "  'Nostoc sp. PCC 7107',\n",
       "  'Pleurocapsa sp. PCC 7327',\n",
       "  'Geitlerinema sp. PCC 7407',\n",
       "  'Acaryochloris marina MBIC11017',\n",
       "  'Cylindrospermum stagnale PCC 7417',\n",
       "  'Thalassoporum mexicanum PCC 7367',\n",
       "  'Calothrix sp. PCC 7507',\n",
       "  'Kamptonema sp. PCC 6506',\n",
       "  'Anabaena cylindrica PCC 7122',\n",
       "  'Stanieria cyanosphaera PCC 7437',\n",
       "  'Crinalium epipsammum PCC 9333',\n",
       "  'Xenococcus sp. PCC 7305',\n",
       "  'Coleofasciculus chthonoplastes PCC 7420',\n",
       "  'Allocoleopsis franciscana PCC 7113',\n",
       "  'Gloeocapsa sp. PCC 73106',\n",
       "  'Chroococcidiopsis thermalis PCC 7203',\n",
       "  'Gloeobacter violaceus PCC 7421',\n",
       "  'Leptolyngbyaceae cyanobacterium JSC-12',\n",
       "  'Fischerella thermalis JSC-11',\n",
       "  'Crocosphaera watsonii WH 8501',\n",
       "  'Lyngbya sp. PCC 8106',\n",
       "  'Limnospira maxima CS-328',\n",
       "  'Anabaena sp. 90',\n",
       "  'Crocosphaera subtropica ATCC 51142',\n",
       "  'Rippkaea orientalis PCC 8801',\n",
       "  'Nostoc punctiforme PCC 73102',\n",
       "  'Nostoc sp. PCC 7120 = FACHB-418',\n",
       "  'Gloeobacter kilaueensis JS1',\n",
       "  'Halothece sp. PCC 7418',\n",
       "  'Gloeothece citriformis PCC 7424',\n",
       "  'Gloeothece verrucosa PCC 7822',\n",
       "  'Synechococcus sp. PCC 7335'],\n",
       " 21: ['Sorangium cellulosum So ce56',\n",
       "  'Myxococcus xanthus DK 1622',\n",
       "  'Haliangium ochraceum DSM 14365',\n",
       "  'Plesiocystis pacifica SIR-1'],\n",
       " 22: ['Celeribacter baekdonensis B30',\n",
       "  'Rhodobacteraceae bacterium HTCC2150',\n",
       "  'Oceaniovalibus guishaninsula JLT2003',\n",
       "  'Cereibacter sphaeroides 2.4.1',\n",
       "  'Salipiger bermudensis HTCC2601',\n",
       "  'Dinoroseobacter shibae DFL 12 = DSM 16493',\n",
       "  'alpha proteobacterium HIMB59',\n",
       "  'Rubellimicrobium thermophilum DSM 16684',\n",
       "  'Ruegeria pomeroyi DSS-3',\n",
       "  'Roseobacter denitrificans OCh 114',\n",
       "  'Paracoccus aminophilus JCM 7686',\n",
       "  'Aurantimonas manganoxydans SI85-9A1',\n",
       "  'Litoreibacter arenae DSM 19593',\n",
       "  'Roseobacter sp. SK209-2-6',\n",
       "  'Roseovarius nubinhibens ISM',\n",
       "  'Salipiger mucosus DSM 16094',\n",
       "  'Polymorphum gilvum SL003B-26A1',\n",
       "  'Roseobacter sp. GAI101',\n",
       "  'Rhodobacteraceae bacterium HIMB11',\n",
       "  'Maritimibacter alkaliphilus HTCC2654',\n",
       "  'Yoonia vestfoldensis SKA53',\n",
       "  'Pseudooceanicola batsensis HTCC2597',\n",
       "  'Ketogulonicigenium vulgare WSH-001',\n",
       "  'Fulvimarina pelagi HTCC2506',\n",
       "  'Oceanibaculum indicum P24',\n",
       "  'Jannaschia sp. CCS1',\n",
       "  'SAR116 cluster alpha proteobacterium HIMB100',\n",
       "  'Ruegeria sp. TM1040',\n",
       "  'Candidatus Pelagibacter sp. HTCC7211',\n",
       "  'Octadecabacter arcticus 238',\n",
       "  'Brucella abortus 2308',\n",
       "  'Thalassobium sp. R2A62',\n",
       "  'Lutibaculum baratangense AMV1',\n",
       "  'Hoeflea phototrophica DFL-43',\n",
       "  'Rhodobacter capsulatus SB 1003',\n",
       "  'Rhodobacterales bacterium HTCC2255',\n",
       "  'Roseivivax marinus',\n",
       "  'Rhodobacter ferrooxidans',\n",
       "  'Paracoccus denitrificans PD1222',\n",
       "  'Candidatus Puniceispirillum marinum IMCC1322',\n",
       "  'Roseibium sp. TrichSKD4',\n",
       "  'alpha proteobacterium BAL199',\n",
       "  'Ahrensia sp. R2A130',\n",
       "  'Rhodobacteraceae bacterium HTCC2083',\n",
       "  'Oceanicola granulosus HTCC2516'],\n",
       " 23: ['Rhodopirellula sallentina SM41',\n",
       "  'Rhodopirellula maiorica SM1',\n",
       "  'Pirellula staleyi DSM 6068',\n",
       "  'Lentisphaera araneosa HTCC2155',\n",
       "  'Isosphaera pallida ATCC 43644',\n",
       "  'Rubinisphaera brasiliensis DSM 5305',\n",
       "  'Rhodopirellula baltica SH 1',\n",
       "  'Planctopirus limnophila DSM 3776',\n",
       "  'Singulisphaera acidiphila DSM 18658',\n",
       "  'Opitutaceae bacterium TAV5'],\n",
       " 24: ['Mycoplasmopsis agalactiae PG2',\n",
       "  'Mycoplasma haemocanis str. Illinois',\n",
       "  'Mycoplasmoides gallisepticum str. R(low)',\n",
       "  'Spiroplasma apis B31',\n",
       "  'Mycoplasma mobile 163K',\n",
       "  'Candidatus Phytoplasma mali AT',\n",
       "  'Mycoplasma putrefaciens Mput9231',\n",
       "  'Spiroplasma syrphidicola EA-1',\n",
       "  'Mycoplasmopsis cynos C142',\n",
       "  'Malacoplasma penetrans HF-2',\n",
       "  'Spiroplasma mirum ATCC 29335',\n",
       "  'Mycoplasmopsis anatis 1340',\n",
       "  'Metamycoplasma arthritidis 158L3-1',\n",
       "  'Metamycoplasma alkalescens 14918',\n",
       "  'Mycoplasmopsis columbina SF7',\n",
       "  'Mycoplasmopsis canis UFG4',\n",
       "  'Spiroplasma culicicola AES-1',\n",
       "  'Spiroplasma sabaudiense Ar-1343',\n",
       "  'Metamycoplasma hominis ATCC 23114',\n",
       "  'Mesomycoplasma conjunctivae HRC/581',\n",
       "  'Mycoplasmopsis synoviae 53',\n",
       "  'Mycoplasmopsis bovigenitalium 51080',\n",
       "  'Mycoplasma wenyonii str. Massachusetts',\n",
       "  'Candidatus Mycoplasma haematolamae str. Purdue',\n",
       "  'Mycoplasmopsis pulmonis UAB CTIP',\n",
       "  'Mycoplasma crocodyli MP145',\n",
       "  'Candidatus Phytoplasma australiense',\n",
       "  'Mycoplasmoides genitalium G37',\n",
       "  'Spiroplasma diminutum CUAS-1',\n",
       "  'Mycoplasma parvum str. Indiana',\n",
       "  'Spiroplasma taiwanense CT-1',\n",
       "  'Mesoplasma florum L1',\n",
       "  'Mycoplasma suis str. Illinois',\n",
       "  'Ureaplasma parvum serovar 3 str. ATCC 700970',\n",
       "  'Mesomycoplasma bovoculi M165/69',\n",
       "  'Mycoplasmopsis fermentans PG18'],\n",
       " 25: ['Methyloversatilis universalis FAM5',\n",
       "  'Azoarcus olearius',\n",
       "  'Thauera linaloolentis 47Lol = DSM 12138',\n",
       "  'Thauera sp. 27',\n",
       "  'Methylibium petroleiphilum PM1',\n",
       "  'Rubrivivax gelatinosus IL144',\n",
       "  'Rhodoferax ferrireducens T118',\n",
       "  'Azoarcus sp. KH32C',\n",
       "  'Herminiimonas arsenicoxydans',\n",
       "  'Thauera terpenica 58Eu',\n",
       "  \"Leptospira biflexa serovar Patoc strain 'Patoc 1 (Paris)'\",\n",
       "  'Oxalobacteraceae bacterium IMCC9480',\n",
       "  'Thauera aminoaromatica',\n",
       "  'Leptothrix cholodnii SP-6',\n",
       "  'Polaromonas naphthalenivorans CJ2',\n",
       "  'Burkholderiales bacterium JOSHI_001',\n",
       "  'Leptospira weilii serovar Topaz str. LT2116',\n",
       "  'Aromatoleum aromaticum EbN1'],\n",
       " 26: ['Melioribacter roseus P3M-2',\n",
       "  'Chloracidobacterium thermophilum B',\n",
       "  'Phycisphaera mikurensis NBRC 102666',\n",
       "  'Elusimicrobium minutum Pei191',\n",
       "  'Akkermansia muciniphila ATCC BAA-835',\n",
       "  'Nitrospina gracilis 3/211',\n",
       "  'Saccharicrinis fermentans DSM 9555 = JCM 21142',\n",
       "  'Chthonomonas calidirosea T49',\n",
       "  'Salinibacter ruber DSM 13855',\n",
       "  'Caldithrix abyssi DSM 13497',\n",
       "  'Chloroherpeton thalassium ATCC 35110'],\n",
       " 27: ['Desulfatibacillum aliphaticivorans',\n",
       "  'Pseudodesulfovibrio piezophilus C1TLV30',\n",
       "  'Desulfocapsa sulfexigens DSM 10523',\n",
       "  'Thermodesulfovibrio yellowstonii DSM 11347',\n",
       "  'Solidesulfovibrio carbinoliphilus subsp. oakridgensis',\n",
       "  'Desulfosudis oleivorans Hxd3',\n",
       "  'Desulfobulbus propionicus DSM 2032',\n",
       "  'Desulfotalea psychrophila LSv54',\n",
       "  'Desulfarculus baarsii DSM 2075',\n",
       "  'Geobacter sulfurreducens PCA',\n",
       "  'Oleidesulfovibrio alaskensis G20',\n",
       "  'Citrifermentans bemidjiense Bem',\n",
       "  'Maridesulfovibrio salexigens DSM 2638',\n",
       "  'Desulfomonile tiedjei DSM 6799',\n",
       "  'Desulfurispirillum indicum S5',\n",
       "  'Denitrovibrio acetiphilus DSM 12809',\n",
       "  'Syntrophotalea carbinolica DSM 2380',\n",
       "  'Bilophila wadsworthia 3_1_6',\n",
       "  'Syntrophus aciditrophicus SB',\n",
       "  'Desulfotignum phosphitoxidans DSM 13687',\n",
       "  'Desulfobacula toluolica Tol2',\n",
       "  'delta proteobacterium NaphS2',\n",
       "  'Desulfovibrio sp. X2',\n",
       "  'Geotalea daltonii FRC-32',\n",
       "  'Alkalidesulfovibrio alkalitolerans DSM 16529',\n",
       "  'Desulfococcus multivorans DSM 2059',\n",
       "  'Geobacter metallireducens GS-15',\n",
       "  'Syntrophobacter fumaroxidans MPOB',\n",
       "  'Calditerrivibrio nitroreducens DSM 19672',\n",
       "  'Desulfohalobium retbaense DSM 5692',\n",
       "  'Desulfurivibrio alkaliphilus AHT 2',\n",
       "  'Deferribacter desulfuricans SSM1',\n",
       "  'Thermodesulfatator indicus DSM 15286',\n",
       "  'Pelobacter propionicus DSM 2379',\n",
       "  'Megalodesulfovibrio gigas DSM 1382 = ATCC 19364',\n",
       "  'Geotalea uraniireducens Rf4',\n",
       "  'Flexistipes sinusarabici DSM 4947',\n",
       "  'Desulfocurvibacter africanus subsp. africanus str. Walvis Bay',\n",
       "  'Desulfovibrio sp. 6_1_46AFAA',\n",
       "  'Desulfomicrobium baculatum DSM 4028',\n",
       "  'Trichlorobacter lovleyi SZ',\n",
       "  'Desulfobacca acetoxidans DSM 11109',\n",
       "  'Nitratidesulfovibrio vulgaris str. Hildenborough',\n",
       "  'Desulfobacter postgatei 2ac9',\n",
       "  'Desulforapulum autotrophicum HRM2',\n",
       "  'Desulfuromonas acetoxidans DSM 684',\n",
       "  'Wolinella succinogenes DSM 1740',\n",
       "  'Desulfonatronospira thiodismutans ASO3-1'],\n",
       " 28: ['Leptospirillum ferrodiazotrophum',\n",
       "  'Anaeromyxobacter sp. Fw109-5',\n",
       "  'Turneriella parva DSM 21527',\n",
       "  'Chlorobaculum tepidum TLS',\n",
       "  'Leptospirillum ferrooxidans C2-3',\n",
       "  'Candidatus Jettenia caeni',\n",
       "  'Chlorobium phaeobacteroides DSM 266',\n",
       "  'Leptonema illini DSM 21528',\n",
       "  'Pelodictyon phaeoclathratiforme BU-1',\n",
       "  'Nitrospira defluvii',\n",
       "  'Chlorobium ferrooxidans DSM 13031',\n",
       "  'Pelodictyon luteolum DSM 273',\n",
       "  'Leptospira interrogans serovar Lai str. 56601',\n",
       "  'Anaeromyxobacter dehalogenans 2CP-C'],\n",
       " 29: ['Coraliomargarita akajimensis DSM 45221',\n",
       "  'Gemmatimonas aurantiaca T-27',\n",
       "  'Corallococcus coralloides DSM 2259',\n",
       "  'Verrucomicrobiia bacterium DG1235',\n",
       "  'Pedosphaera parvula Ellin514',\n",
       "  'Opitutus terrae PB90-1',\n",
       "  'Chthoniobacter flavus Ellin428'],\n",
       " 30: ['Adlercreutzia caecimuris B7',\n",
       "  'Eggerthella lenta DSM 2243',\n",
       "  'Desulfoscipio gibsoniae DSM 7213',\n",
       "  'Desulforamulus reducens MI-1',\n",
       "  'Alkaliphilus metalliredigens QYMF',\n",
       "  'Desulfitobacterium hafniense Y51',\n",
       "  'Desulfitobacterium metallireducens DSM 15288',\n",
       "  'Syntrophomonas wolfei subsp. wolfei str. Goettingen G311',\n",
       "  'Desulfitobacterium dichloroeliminans LMG P-21439',\n",
       "  'Slackia heliotrinireducens DSM 20476',\n",
       "  'Desulfofarcimen acetoxidans DSM 771',\n",
       "  'Thermincola potens JR',\n",
       "  'Desulforamulus hydrothermalis Lam5 = DSM 18033',\n",
       "  'Desulforamulus ruminis DSM 2154',\n",
       "  'Desulfotomaculum nigrificans CO-1-SRB',\n",
       "  'Alkaliphilus oremlandii OhILAs',\n",
       "  'Eggerthella sp. YY7918',\n",
       "  'Ammonifex degensii KC4',\n",
       "  'Carboxydothermus hydrogenoformans Z-2901',\n",
       "  'Gordonibacter pamelaeae 7-10-1-b',\n",
       "  'Heliomicrobium modesticaldum Ice1',\n",
       "  'Candidatus Desulforudis audaxviator MP104C',\n",
       "  'Acetonema longum DSM 6540',\n",
       "  'Syntrophobotulus glycolicus DSM 8271',\n",
       "  'Pelotomaculum thermopropionicum SI',\n",
       "  'Dethiobacter alkaliphilus AHT 1',\n",
       "  'Desulfosporosinus orientis DSM 765',\n",
       "  'Slackia piriformis YIT 12062',\n",
       "  'Desulfosporosinus youngiae DSM 17734',\n",
       "  'Slackia exigua ATCC 700122',\n",
       "  'Thermosinus carboxydivorans Nor1',\n",
       "  'Dehalococcoides mccartyi VS',\n",
       "  'Thermacetogenium phaeum DSM 12270',\n",
       "  'Syntrophothermus lipocalidus DSM 12680',\n",
       "  'Symbiobacterium thermophilum IAM 14863'],\n",
       " 31: ['Methylorubrum extorquens AM1',\n",
       "  'Hyphomicrobium denitrificans ATCC 51888',\n",
       "  'Hyphomicrobium nitrativorans NL23',\n",
       "  'Commensalibacter papalotli (ex Servin-Garciduenas et al. 2014)',\n",
       "  'Rhodospirillum rubrum ATCC 11170',\n",
       "  'Acetobacter pomorum DM001',\n",
       "  'Rhodospirillum centenum SW',\n",
       "  'Gluconobacter oxydans 621H',\n",
       "  'Nitrobacter hamburgensis X14',\n",
       "  'Rhodomicrobium vannielii ATCC 17100',\n",
       "  'Nitrobacter winogradskyi Nb-255',\n",
       "  'Afipia carboxidovorans OM5',\n",
       "  'Granulibacter bethesdensis CGDNIH1',\n",
       "  'Gluconobacter morbifer G707',\n",
       "  'Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821',\n",
       "  'Magnetospirillum gryphiswaldense MSR-1',\n",
       "  'Methylocella silvestris BL2',\n",
       "  'Nitrobacter sp. Nb-311A',\n",
       "  'Beijerinckia indica subsp. indica ATCC 9039',\n",
       "  'Caenispirillum salinarum AK4',\n",
       "  'Acidocella sp. MX-AZ02',\n",
       "  'Hyphomicrobium sp. MC1',\n",
       "  'Rhodovulum sp. PH10',\n",
       "  'Bartonella tamiae Th239',\n",
       "  'Acidiphilium cryptum JF-5',\n",
       "  'Gluconacetobacter diazotrophicus PA1 5',\n",
       "  'Parvularcula bermudensis HTCC2503'],\n",
       " 32: ['Treponema pedis str. T A4',\n",
       "  'Treponema pallidum subsp. pallidum str. Nichols',\n",
       "  'Leadbettera azotonutricia ZAS-9',\n",
       "  'Treponema primitia ZAS-2',\n",
       "  'Brachyspira pilosicoli 95/1000',\n",
       "  'Treponema denticola ATCC 35405',\n",
       "  'Salinispira pacifica',\n",
       "  'Treponema saccharophilum DSM 2985',\n",
       "  'Gracilinema caldarium DSM 7334',\n",
       "  'Parasphaerochaeta coccoides DSM 17374',\n",
       "  'Winmispira thermophila DSM 6578',\n",
       "  'Treponema succinifaciens DSM 2489',\n",
       "  'Treponema vincentii F0403',\n",
       "  'Treponema maltophilum ATCC 51939',\n",
       "  'Sphaerochaeta globosa str. Buddy',\n",
       "  'Treponema sp. JC4',\n",
       "  'Sediminispirochaeta smaragdinae DSM 11293',\n",
       "  'Spirochaeta africana DSM 8902',\n",
       "  'Sphaerochaeta pleomorpha str. Grapes',\n",
       "  'Treponema brennaborense DSM 12168'],\n",
       " 33: ['Escherichia coli O139:H28 str. E24377A',\n",
       "  'Escherichia coli APEC O1',\n",
       "  'Yersinia pestis',\n",
       "  'Photorhabdus laumondii subsp. laumondii TTO1',\n",
       "  'Klebsiella pneumoniae subsp. pneumoniae HS11286',\n",
       "  '[Enterobacter] lignolyticus SCF1',\n",
       "  'Enterobacteriaceae bacterium strain FGI 57',\n",
       "  'Dickeya dadantii 3937',\n",
       "  'Rahnella aquatilis CIP 78.65 = ATCC 33071',\n",
       "  'Escherichia coli O127:H6 str. E2348/69',\n",
       "  'Shigella sonnei Ss046',\n",
       "  'Shigella boydii CDC 3083-94',\n",
       "  'Serratia sp. DD3',\n",
       "  'Cronobacter sakazakii ATCC BAA-894',\n",
       "  'Shigella dysenteriae Sd197',\n",
       "  'Shimwellia blattae DSM 4481 = NBRC 105725',\n",
       "  'Escherichia coli O157:H7',\n",
       "  'Salmonella enterica subsp. enterica serovar Typhimurium str. LT2',\n",
       "  'Pantoea ananatis LMG 20103',\n",
       "  'Pectobacterium atrosepticum SCRI1043',\n",
       "  'Shigella flexneri',\n",
       "  'Xenorhabdus nematophila ATCC 19061',\n",
       "  'Citrobacter koseri ATCC BAA-895',\n",
       "  'Candidatus Sodalis pierantonii str. SOPE',\n",
       "  'Sodalis praecaptivus',\n",
       "  'Erwinia billingiae Eb661',\n",
       "  'Erwinia tasmaniensis Et1/99',\n",
       "  'Escherichia coli 55989',\n",
       "  'Escherichia coli S88',\n",
       "  'Xenorhabdus szentirmaii DSM 16338',\n",
       "  'Plautia stali symbiont',\n",
       "  'Morganella morganii subsp. morganii KT',\n",
       "  'Serratia odorifera DSM 4582',\n",
       "  'Salmonella enterica subsp. arizonae serovar 62:z4,z23:-',\n",
       "  'Atlantibacter hermannii NBRC 105704',\n",
       "  'Prodigiosinella confusarubida',\n",
       "  'Escherichia coli CFT073',\n",
       "  'Escherichia coli K-12',\n",
       "  'Escherichia coli O83:H1 str. NRG 857C',\n",
       "  'Providencia sneebia DSM 19967'],\n",
       " 34: ['[Clostridium] innocuum 2959',\n",
       "  'Thomasclavelia ramosa DSM 1402',\n",
       "  'Clostridium kluyveri DSM 555',\n",
       "  'Pelosinus fermentans JBW45',\n",
       "  'Acetobacterium woodii DSM 1030',\n",
       "  'Thomasclavelia spiroformis DSM 1552',\n",
       "  'Clostridium sp. DL-VIII',\n",
       "  'Turicibacter sp. HGF1',\n",
       "  'Clostridium novyi NT',\n",
       "  'Clostridium tetani E88',\n",
       "  'Clostridium acetobutylicum ATCC 824',\n",
       "  'Ethanoligenens harbinense YUAN-3',\n",
       "  'Clostridium celatum DSM 1785',\n",
       "  'Anaerofustis stercorihominis DSM 17244',\n",
       "  'Clostridium butyricum E4 str. BoNT E BL5262',\n",
       "  'Clostridium perfringens str. 13',\n",
       "  'Clostridium thermobutyricum',\n",
       "  'Clostridium sp. 7_2_43FAA',\n",
       "  'Clostridium sp. BL8',\n",
       "  'Clostridium carboxidivorans P7',\n",
       "  'Erysipelotrichaceae bacterium 5_2_54FAA',\n",
       "  'Clostridium intestinale URNW',\n",
       "  'Clostridium sp. MSTE9',\n",
       "  'Clostridium botulinum A str. Hall',\n",
       "  'Coprobacillus cateniformis',\n",
       "  'Gottschalkia acidurici 9a',\n",
       "  'Clostridiales bacterium oral taxon 876 str. F0540',\n",
       "  'Clostridium pasteurianum BC1',\n",
       "  'Clostridium saccharobutylicum DSM 13864',\n",
       "  'Clostridium saccharoperbutylacetonicum N1-4(HMT)',\n",
       "  'Clostridioides difficile 630',\n",
       "  'Haloplasma contractile SSD-17B'],\n",
       " 35: ['Actinomyces massiliensis F0489',\n",
       "  'Corynebacterium halotolerans YIM 70093 = DSM 44683',\n",
       "  'Corynebacterium urealyticum DSM 7109',\n",
       "  'Corynebacterium argentoratense DSM 44202',\n",
       "  'Micrococcus luteus NCTC 2665',\n",
       "  'Corynebacterium sp. KPL1989',\n",
       "  'Corynebacterium sp. KPL1821',\n",
       "  'Rothia mucilaginosa DY-18',\n",
       "  'Jonesia denitrificans DSM 20603',\n",
       "  'Bifidobacterium animalis subsp. lactis AD011',\n",
       "  'Corynebacterium glutamicum ATCC 13032',\n",
       "  'Actinomyces urogenitalis DSM 15434',\n",
       "  'Bifidobacterium bifidum BGN4',\n",
       "  'Corynebacterium otitidis ATCC 51513',\n",
       "  'Winkia neuii BV029A5',\n",
       "  'Kocuria rhizophila DC2201',\n",
       "  'Parascardovia denticolens DSM 10105 = JCM 12538',\n",
       "  'Corynebacterium durum F0235',\n",
       "  'Corynebacterium pseudotuberculosis C231',\n",
       "  'Bifidobacterium adolescentis ATCC 15703',\n",
       "  'Leifsonia xyli subsp. xyli str. CTCB07',\n",
       "  'Actinomyces sp. oral taxon 849 str. F0330',\n",
       "  'Actinotignum schaalii FB123-CNA-2',\n",
       "  'Corynebacterium pyruviciproducens ATCC BAA-1742',\n",
       "  'Propionimicrobium lymphophilum ACS-093-V-SCH5',\n",
       "  'Corynebacterium efficiens YS-314',\n",
       "  'Propionibacterium sp. oral taxon 192 str. F0372',\n",
       "  'Actinomyces graevenitzii C83',\n",
       "  'Dermabacter sp. HFH0086',\n",
       "  'Corynebacterium matruchotii ATCC 14266',\n",
       "  'Corynebacterium kroppenstedtii DSM 44385',\n",
       "  'Schaalia turicensis ACS-279-V-Col4',\n",
       "  'Schaalia cardiffensis F0333',\n",
       "  'Propionibacterium freudenreichii subsp. shermanii CIRM-BIA1',\n",
       "  'Mobiluncus mulieris ATCC 35239',\n",
       "  'Cutibacterium granulosum DSM 20700',\n",
       "  'Mobiluncus curtisii ATCC 43063',\n",
       "  'Corynebacterium genitalium ATCC 33030',\n",
       "  'Actinobaculum sp. oral taxon 183 str. F0552',\n",
       "  'Alloscardovia omnicolens F0580',\n",
       "  'Corynebacterium maris DSM 45190',\n",
       "  'Kocuria palustris PEL',\n",
       "  'Bifidobacterium longum NCC2705',\n",
       "  'Arcanobacterium haemolyticum DSM 20595',\n",
       "  'Corynebacterium jeikeium K411',\n",
       "  'Gleimia coleocanis DSM 15436',\n",
       "  'Scardovia inopinata F0304',\n",
       "  'Tropheryma whipplei str. Twist',\n",
       "  'Actinomyces sp. oral taxon 448 str. F0400',\n",
       "  'Cutibacterium avidum ATCC 25577',\n",
       "  'Corynebacterium resistens DSM 45100',\n",
       "  'Brevibacterium mcbrellneri ATCC 49030',\n",
       "  'Corynebacterium aurimucosum ATCC 700975',\n",
       "  'Corynebacterium vitaeruminis DSM 20294',\n",
       "  'Corynebacterium diphtheriae NCTC 13129',\n",
       "  'Scardovia wiggsiae F0424',\n",
       "  'Actinobaculum massiliense ACS-171-V-Col2'],\n",
       " 36: ['Thermus thermophilus HB8',\n",
       "  'Truepera radiovictrix DSM 17093',\n",
       "  'Roseiflexus castenholzii DSM 13941',\n",
       "  'Caldilinea aerophila DSM 14535 = NBRC 104270',\n",
       "  'Deinococcus maricopensis DSM 21211',\n",
       "  'Deinococcus peraridilitoris DSM 19664',\n",
       "  'Deinococcus radiodurans R1 = ATCC 13939 = DSM 20539',\n",
       "  'Sphaerobacter thermophilus DSM 20745',\n",
       "  'Marinithermus hydrothermalis DSM 14884',\n",
       "  'Rubrobacter xylanophilus DSM 9941',\n",
       "  'Anaerolinea thermophila UNI-1',\n",
       "  'Oceanithermus profundus DSM 14977',\n",
       "  'Deinococcus proteolyticus MRP',\n",
       "  'Chloroflexus aurantiacus J-10-fl',\n",
       "  'Allomeiothermus silvanus DSM 9946'],\n",
       " 37: ['Sphingobium indicum UT26S',\n",
       "  'Sphingobium ummariense RL-3',\n",
       "  'Sphingobium baderi LL03',\n",
       "  'Novosphingobium lindaniclasticum LE124',\n",
       "  'Novosphingobium aromaticivorans DSM 12444',\n",
       "  'Rhizorhabdus wittichii RW1',\n",
       "  'Novosphingobium pentaromativorans US6-1',\n",
       "  'Sphingomonas sanxanigenens DSM 19645 = NX02',\n",
       "  'Sphingobium sp. C100',\n",
       "  'Sphingobium indicum BiD32',\n",
       "  'Sphingobium chlorophenolicum L-1',\n",
       "  'Brevundimonas abyssalis TAR-001',\n",
       "  'Erythrobacter litoralis HTCC2594',\n",
       "  'Novosphingobium sp. Rr 2-17',\n",
       "  'Novosphingobium nitrogenifigens DSM 19370',\n",
       "  'Phenylobacterium zucineum HLK1',\n",
       "  'Parvibaculum lavamentivorans DS-1',\n",
       "  'Pacificimonas flava',\n",
       "  'Erythrobacter sp. NAP1',\n",
       "  'Sphingopyxis alaskensis RB2256',\n",
       "  'Sphingomonas sp. MM-1',\n",
       "  'Caenibius tardaugens NBRC 16725',\n",
       "  'Sphingobium yanoikuyae ATCC 51230',\n",
       "  'Sphingobium sp. SYK-6',\n",
       "  'Novosphingobium sp. AP12',\n",
       "  'Novosphingobium sp. PP1Y',\n",
       "  'Sphingomonas sp. SKA58'],\n",
       " 38: ['Acidaminococcus fermentans DSM 20731',\n",
       "  'Leptotrichia sp. oral taxon 215 str. W9775',\n",
       "  'Acidaminococcus intestini RyC-MR95',\n",
       "  'Dialister micraerophilus DSM 19965',\n",
       "  'Streptobacillus moniliformis DSM 12112',\n",
       "  'Mitsuokella sp. oral taxon 131 str. W9106',\n",
       "  'Phascolarctobacterium succinatutens YIT 12067',\n",
       "  'Pseudoleptotrichia goodfellowii F0264',\n",
       "  'Leptotrichia hofstadii F0254',\n",
       "  'Fusobacterium gonidiaformans 3-1-5R',\n",
       "  'Veillonella sp. HPA0037',\n",
       "  'Cetobacterium somerae ATCC BAA-474',\n",
       "  'Megasphaera vaginalis (ex Srinivasan et al. 2021)',\n",
       "  'Megasphaera elsdenii DSM 20460',\n",
       "  'Youngiibacter fragilis 232.1',\n",
       "  'Anaeroglobus geminatus F0357',\n",
       "  'Fusobacterium nucleatum subsp. nucleatum ATCC 25586',\n",
       "  'Veillonella sp. oral taxon 780 str. F0422',\n",
       "  'Sebaldella termitidis ATCC 33386'],\n",
       " 39: ['Megasphaera micronuciformis F0359',\n",
       "  'Tepidanaerobacter acetatoxydans Re1',\n",
       "  'Dialister succinatiphilus YIT 11850',\n",
       "  'Thermosediminibacter oceani DSM 16646',\n",
       "  'Halobacteroides halobius DSM 5150',\n",
       "  'Caldisalinibacter kiritimatiensis',\n",
       "  'Selenomonas sputigena ATCC 35185',\n",
       "  'Acetoanaerobium sticklandii DSM 519',\n",
       "  'Caldanaerobacter subterraneus subsp. tengcongensis MB4',\n",
       "  'Thermoanaerobacter pseudethanolicus ATCC 33223',\n",
       "  'Caloramator australicus RC3',\n",
       "  'Thermoanaerobacterium thermosaccharolyticum DSM 571',\n",
       "  'Acetohalobium arabaticum DSM 5501',\n",
       "  'Natranaerobius thermophilus JW/NM-WN-LF',\n",
       "  'Selenomonas sp. F0473',\n",
       "  'Thermobrachium celere DSM 8682'],\n",
       " 40: ['Sulfobacillus acidophilus DSM 10332',\n",
       "  'Thermobaculum terrenum ATCC BAA-798',\n",
       "  'Ktedonobacter racemifer DSM 44963'],\n",
       " 41: ['Lentilactobacillus buchneri subsp. silagei CD034',\n",
       "  'Loigolactobacillus coryniformis subsp. coryniformis CECT 5711',\n",
       "  'Fructilactobacillus sanfranciscensis TMW 1.1304',\n",
       "  'Ligilactobacillus ruminis ATCC 27782',\n",
       "  'Lactobacillus delbrueckii subsp. bulgaricus ATCC 11842 = JCM 1002',\n",
       "  'Lentilactobacillus hilgardii DSM 20176 = ATCC 8290',\n",
       "  'Streptococcus thermophilus LMG 18311',\n",
       "  'Streptococcus gordonii str. Challis substr. CH1',\n",
       "  'Limosilactobacillus coleohominis 101-4-CHN',\n",
       "  'Lactobacillus amylolyticus DSM 11664',\n",
       "  'Lactococcus cremoris subsp. cremoris TIFN6',\n",
       "  'Enterococcus sulfureus ATCC 49903',\n",
       "  'Lactobacillus iners DSM 13335',\n",
       "  'Catellicoccus marimammalium M35/04/3',\n",
       "  'Streptococcus intermedius B196',\n",
       "  'Vagococcus lutrae LBD1',\n",
       "  'Facklamia languida CCUG 37842',\n",
       "  'Streptococcus pneumoniae TIGR4',\n",
       "  'Streptococcus pneumoniae R6',\n",
       "  'Leuconostoc carnosum JB16',\n",
       "  'Lentilactobacillus parafarraginis F0439',\n",
       "  'Enterococcus cecorum DSM 20682 = ATCC 43198',\n",
       "  'Limosilactobacillus gastricus PS3',\n",
       "  'Streptococcus agalactiae 2603V/R',\n",
       "  'Lactobacillus acidophilus NCFM',\n",
       "  'Pediococcus acidilactici DSM 20284',\n",
       "  'Facklamia hominis CCUG 36813',\n",
       "  'Streptococcus pyogenes serotype M1',\n",
       "  'Oenococcus oeni PSU-1',\n",
       "  'Mycoplasmoides pneumoniae M129',\n",
       "  'Ligilactobacillus salivarius UCC118',\n",
       "  'Listeria monocytogenes EGD-e',\n",
       "  'Weissella paramesenteroides ATCC 33313',\n",
       "  'Lactobacillus pasteurii DSM 23907 = CRBIP 24.76',\n",
       "  'Alloiococcus otitis ATCC 51267',\n",
       "  'Streptococcus suis BM407',\n",
       "  'Lactococcus lactis subsp. lactis Il1403',\n",
       "  'Listeria grayi DSM 20601',\n",
       "  'Enterococcus italicus DSM 15952',\n",
       "  'Streptococcus sanguinis SK36',\n",
       "  'Streptococcus mutans UA159',\n",
       "  'Limosilactobacillus reuteri subsp. reuteri',\n",
       "  'Latilactobacillus sakei subsp. sakei 23K',\n",
       "  'Lactococcus garvieae Lg2',\n",
       "  'Melissococcus plutonius ATCC 35311',\n",
       "  'Streptococcus infantis ATCC 700779',\n",
       "  'Granulicatella elegans ATCC 700633',\n",
       "  'Carnobacterium sp. 17-4',\n",
       "  'Limosilactobacillus mucosae LM1',\n",
       "  'Enterococcus saccharolyticus subsp. saccharolyticus ATCC 43076',\n",
       "  'Enterococcus dispar ATCC 51266',\n",
       "  'Dolosigranulum pigrum ATCC 51524',\n",
       "  'Enterococcus columbae DSM 7374 = ATCC 51263',\n",
       "  'Enterococcus faecalis V583',\n",
       "  'Enterococcus phoeniculicola ATCC BAA-412',\n",
       "  'Enterococcus pallens ATCC BAA-351',\n",
       "  'Levilactobacillus brevis ATCC 367',\n",
       "  'Streptococcus ictaluri 707-05',\n",
       "  'Streptococcus pneumoniae D39',\n",
       "  'Enterococcus casseliflavus EC20',\n",
       "  'Streptococcus pasteurianus ATCC 43144',\n",
       "  'Ligilactobacillus saerimneri 30a',\n",
       "  'Aerococcus sp. Group 1',\n",
       "  'Erysipelothrix rhusiopathiae ATCC 19414',\n",
       "  'Lacticaseibacillus paracasei ATCC 334',\n",
       "  'Enterococcus asini ATCC 700915',\n",
       "  'Streptococcus uberis 0140J',\n",
       "  'Eremococcus coleocola ACS-139-V-Col8',\n",
       "  'Pediococcus claussenii ATCC BAA-344',\n",
       "  'Brochothrix campestris FSL F6-1037',\n",
       "  'Carnobacterium maltaromaticum LMA28',\n",
       "  'Granulicatella adiacens ATCC 49175',\n",
       "  'Oenococcus kitaharae DSM 17330',\n",
       "  'Lactiplantibacillus plantarum WCFS1',\n",
       "  'Falseniella ignava CCUG 37419',\n",
       "  'Abiotrophia defectiva ATCC 49176',\n",
       "  'Streptococcus urinalis 2285-97',\n",
       "  'Ligilactobacillus equi DPC 6820'],\n",
       " 42: ['Metapseudomonas resinovorans NBRC 106553',\n",
       "  'Pseudomonas sp. HPB0071',\n",
       "  'Pseudomonas syringae pv. tomato str. DC3000',\n",
       "  'Marinomonas mediterranea MMB-1',\n",
       "  'Stutzerimonas stutzeri A1501',\n",
       "  'Pseudogulbenkiania sp. NH8B',\n",
       "  'Hahella chejuensis KCTC 2396',\n",
       "  'Pseudomonas sp. M47T1',\n",
       "  'Pseudomonas fulva 12-X',\n",
       "  'Chromobacterium violaceum ATCC 12472',\n",
       "  'Pseudomonas putida KT2440',\n",
       "  'Pseudomonas aeruginosa PAO1'],\n",
       " 43: ['Granulicella mallensis MP5ACTX8',\n",
       "  'Terriglobus saanensis SP1PR4',\n",
       "  'Gemmatirosa kalamazoonensis',\n",
       "  'Candidatus Koribacter versatilis Ellin345',\n",
       "  'Acidobacterium capsulatum ATCC 51196',\n",
       "  'Terriglobus roseus DSM 18391',\n",
       "  'Nitrolancea hollandica Lb',\n",
       "  'Granulicella tundricola MP5ACTX9'],\n",
       " 44: ['Firmicutes bacterium M10-2',\n",
       "  'Atopobium minutum 10063974',\n",
       "  'candidate division TM7 genomosp. GTL1',\n",
       "  'Fannyhessea vaginae DSM 15829',\n",
       "  'Collinsella tanakaei YIT 12063',\n",
       "  'Coriobacteriaceae bacterium BV3Ac1',\n",
       "  'Candidatus Saccharibacteria bacterium RAAC3_TM7_1',\n",
       "  'Olsenella profusa F0195',\n",
       "  'Collinsella stercoris DSM 13279',\n",
       "  'Mycoplasma mycoides subsp. mycoides SC str. PG1',\n",
       "  'Solobacterium moorei F0204',\n",
       "  'Candidatus Saccharimonas aalborgensis',\n",
       "  'Lancefieldella parvula DSM 20469',\n",
       "  'Cryptobacterium curtum DSM 15641',\n",
       "  'Coriobacterium glomerans PW2',\n",
       "  'Bulleidia extructa W1219',\n",
       "  'Helcococcus kunzii ATCC 51366',\n",
       "  'Olsenella uli DSM 7084',\n",
       "  'Atopobium sp. oral taxon 199 str. F0494',\n",
       "  'Eggerthia catenaformis OT 569 = DSM 20559',\n",
       "  'Olsenella sp. oral taxon 809 str. F0356'],\n",
       " 45: ['Anaerococcus prevotii ACS-065-V-Col13',\n",
       "  'Anaerococcus lactolyticus ATCC 51172',\n",
       "  'Peptoniphilus indolicus ATCC 29427',\n",
       "  'Peptoanaerobacter stomatis',\n",
       "  'Peptoniphilus duerdenii ATCC BAA-1640',\n",
       "  'Peptostreptococcus stomatis DSM 17678',\n",
       "  'Eubacterium nodatum ATCC 33099',\n",
       "  'Peptacetobacter hiranonis DSM 13275',\n",
       "  'Clostridium sp. SY8519',\n",
       "  'Mogibacterium sp. CM50',\n",
       "  'Shuttleworthella satelles DSM 14600',\n",
       "  'Mageeibacillus indolicus UPII9-5',\n",
       "  'Peptostreptococcaceae bacterium oral taxon 113 str. W5053',\n",
       "  'Oribacterium sp. oral taxon 078 str. F0263',\n",
       "  'Pseudoramibacter alactolyticus ATCC 23263',\n",
       "  'Johnsonella ignava ATCC 51276',\n",
       "  'Peptoniphilus sp. oral taxon 386 str. F0131',\n",
       "  'Eubacterium infirmum F0142',\n",
       "  'Peptostreptococcaceae bacterium AS15',\n",
       "  'Peptoniphilus sp. BV3C26',\n",
       "  'Eubacterium brachy ATCC 33089',\n",
       "  'Eubacterium saphenum ATCC 49989',\n",
       "  'Peptoniphilus lacrimalis 315-B',\n",
       "  'Anaerococcus vaginalis ATCC 51170',\n",
       "  'Finegoldia magna ATCC 29328',\n",
       "  'Filifactor alocis ATCC 35896',\n",
       "  'Acholeplasma laidlawii PG-8A',\n",
       "  'Peptoniphilus harei ACS-146-V-Sch2b'],\n",
       " 46: ['Fervidobacterium nodosum Rt17-B1',\n",
       "  'Caldisericum exile AZM16c01',\n",
       "  'Caldicellulosiruptor owensensis OL',\n",
       "  'Kosmotoga olearia TBF 19.5.1',\n",
       "  'Dictyoglomus turgidum DSM 6724',\n",
       "  'Marinitoga piezophila KA3',\n",
       "  'Pseudothermotoga lettingae TMO',\n",
       "  'Mesotoga prima MesG1.Ag.4.2',\n",
       "  'Coprothermobacter proteolyticus DSM 5265',\n",
       "  'Thermotoga maritima MSB8',\n",
       "  'Halothermothrix orenii H 168'],\n",
       " 47: ['Rickettsia prowazekii str. Madrid E',\n",
       "  'Candidatus Blochmanniella pennsylvanica str. BPEN',\n",
       "  'Francisella tularensis subsp. tularensis SCHU S4',\n",
       "  'Eikenella corrodens CC92I',\n",
       "  'Candidatus Regiella insecticola 5.15',\n",
       "  'Serratia symbiotica str. Tucson',\n",
       "  'Candidatus Profftella armatura',\n",
       "  'Actinobacillus pleuropneumoniae serovar 5b str. L20',\n",
       "  'Oxalobacter formigenes OXCC13',\n",
       "  'Candidatus Carsonella ruddii HC isolate Thao2000',\n",
       "  'secondary endosymbiont of Ctenarytaina eucalypti',\n",
       "  'Neisseria sp. oral taxon 020 str. F0370',\n",
       "  'Dichelobacter nodosus VCS1703A',\n",
       "  'Candidatus Vesicomyidisocius calyptogenae',\n",
       "  'Buchnera aphidicola BCc',\n",
       "  'Actinobacillus succinogenes 130Z',\n",
       "  'Glaesserella parasuis SH0165',\n",
       "  'Sutterella parvirubra YIT 11816',\n",
       "  'Succinatimonas hippei YIT 12066',\n",
       "  'Kingella kingae ATCC 23330',\n",
       "  'Buchnera aphidicola str. Bp (Baizongia pistaciae)',\n",
       "  'Simonsiella muelleri ATCC 29453',\n",
       "  'Kingella denitrificans ATCC 33394',\n",
       "  '[Haemophilus] ducreyi 35000HP',\n",
       "  'Cardiobacterium hominis ATCC 15826',\n",
       "  'Wigglesworthia glossinidia endosymbiont of Glossina brevipalpis',\n",
       "  'Buchnera aphidicola (Cinara tujafilina)',\n",
       "  'Neisseria bacilliformis ATCC BAA-1200',\n",
       "  'Haemophilus influenzae Rd KW20',\n",
       "  'secondary endosymbiont of Heteropsylla cubana',\n",
       "  'Wolbachia endosymbiont of Onchocerca ochengi',\n",
       "  'Pasteurella bettyae CCUG 2042',\n",
       "  'Sutterella wadsworthensis HGA0223',\n",
       "  'Taylorella asinigenitalis MCE3',\n",
       "  'Thiomicrospira cyclica ALM1',\n",
       "  'Candidatus Regiella insecticola LSR1',\n",
       "  'Kingella oralis ATCC 51147',\n",
       "  'Pasteurella multocida subsp. multocida str. Pm70',\n",
       "  'Anaplasma marginale str. Florida',\n",
       "  'Candidatus Pantoea carbekii',\n",
       "  'Candidatus Zinderia insecticola CARI',\n",
       "  'Candidatus Kinetoplastidibacterium crithidiae TCC036E',\n",
       "  'Baumannia cicadellinicola str. Hc (Homalodisca coagulata)',\n",
       "  'Blattabacterium sp. (Periplaneta americana) str. BPLAN',\n",
       "  'Candidatus Moranella endobia PCIT',\n",
       "  'Neisseria shayeganii 871',\n",
       "  'Sutterella wadsworthensis 2_1_59BFAA',\n",
       "  'Ehrlichia chaffeensis str. Arkansas',\n",
       "  'Wohlfahrtiimonas chitiniclastica SH04',\n",
       "  'Candidatus Kinetoplastidibacterium blastocrithidiae TCC012E',\n",
       "  'Neisseria meningitidis MC58',\n",
       "  'Candidatus Uzinura diaspidicola str. ASNER',\n",
       "  'Buchnera aphidicola str. APS (Acyrthosiphon pisum)',\n",
       "  'Neisseria gonorrhoeae FA 1090']}"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters_names = {}\n",
    "for i in range(n_clusters):\n",
    "\tclusters_names[i] = [tid2nm[tid] if tid in tid2nm else 'unknown' for tid in clusters[i]]\n",
    "\n",
    "clusters_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "650a7cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def index_tree(tree_dir: nx.DiGraph):\n",
    "    # Find root (in-degree 0)\n",
    "    roots = [v for v in tree_dir.nodes if tree_dir.in_degree(v) == 0]\n",
    "    if len(roots) != 1:\n",
    "        raise ValueError(f\"Expected exactly one root, found {len(roots)}\")\n",
    "    root = roots[0]\n",
    "\n",
    "    # Build parent/children\n",
    "    parent = {root: None}\n",
    "    children = {v: [] for v in tree_dir.nodes}\n",
    "    for u, v in tree_dir.edges:\n",
    "        parent[v] = u\n",
    "        children[u].append(v)\n",
    "\n",
    "    # Iterative DFS to get tin, depth, subtree_size\n",
    "    tin, depth, subtree_size = {}, {}, {}\n",
    "    node_by_tin = []\n",
    "    t = 0\n",
    "    depth[root] = 0\n",
    "    stack = [(root, 0)]  # (node, 0=enter, 1=exit)\n",
    "    while stack:\n",
    "        v, state = stack.pop()\n",
    "        if state == 0:\n",
    "            tin[v] = t\n",
    "            node_by_tin.append(v)\n",
    "            t += 1\n",
    "            stack.append((v, 1))\n",
    "            # Push children in reverse so leftmost visits first\n",
    "            for w in reversed(children[v]):\n",
    "                depth[w] = depth[v] + 1\n",
    "                stack.append((w, 0))\n",
    "        else:\n",
    "            s = 1\n",
    "            for w in children[v]:\n",
    "                s += subtree_size[w]\n",
    "            subtree_size[v] = s\n",
    "\n",
    "    tout = {v: tin[v] + subtree_size[v] - 1 for v in tree_dir.nodes}\n",
    "    return {\n",
    "        \"root\": root, \"parent\": parent, \"children\": children, \"depth\": depth,\n",
    "        \"tin\": tin, \"tout\": tout, \"subtree_size\": subtree_size, \"n\": len(node_by_tin)\n",
    "    }\n",
    "\n",
    "def _normalize(u):\n",
    "    # e.g. 795666.0 -> 795666\n",
    "    return int(u) if isinstance(u, float) and u.is_integer() else u\n",
    "\n",
    "def cluster_tree_stats(idx, cluster_nodes, alpha=0.8, size_cap=None):\n",
    "    \"\"\"\n",
    "    Compute robust clade-based summaries for a cluster:\n",
    "      - best coverage clade (max |C∩subtree(v)| / |C|)\n",
    "      - best Jaccard clade ( |C∩S| / (|C| + |S| - |C∩S|) ) to penalize huge clades\n",
    "      - alpha-majority clade: deepest v with |C∩subtree(v)| >= ceil(alpha*|C|)\n",
    "    Optional: size_cap limits candidate clades to subtree_size(v) <= size_cap.\n",
    "    \"\"\"\n",
    "\n",
    "    tin, tout, depth, parent, subtree_size = (idx[\"tin\"], idx[\"tout\"], idx[\"depth\"], idx[\"parent\"], idx[\"subtree_size\"])\n",
    "\n",
    "    C_all = list(cluster_nodes)\t# list \n",
    "    C = [_normalize(u) for u in C_all if u in tin]  # filter to nodes present in this tree\n",
    "    skipped = len(C_all) - len(C)\n",
    "    if len(C) == 0:\n",
    "        return {\"n_in_cluster\": 0, \"n_skipped\": skipped}\n",
    "\n",
    "    # Pre-sort cluster members in preorder\n",
    "    C_tins = sorted(tin[u] for u in C)\n",
    "\n",
    "    def count_in_subtree(v):\n",
    "        a, b = tin[v], tout[v]\n",
    "        return bisect_right(C_tins, b) - bisect_left(C_tins, a)\n",
    "\n",
    "    # Candidates = union of ancestors of the cluster (small)\n",
    "    candidates = set()\n",
    "    for u in C:\n",
    "        v = u\n",
    "        while v is not None and v not in candidates:\n",
    "            if size_cap is None or subtree_size[v] <= size_cap:\n",
    "                candidates.add(v)\n",
    "            v = parent[v]\n",
    "\n",
    "    best_cov_node, best_cov, best_cov_depth = None, -1, -1\n",
    "    best_j_node, best_j, best_j_depth = None, -1.0, -1\n",
    "    thr = math.ceil(alpha * len(C))\n",
    "    alpha_node, alpha_cov, alpha_depth = None, 0, -1\n",
    "\n",
    "    for v in candidates:\n",
    "        cov = count_in_subtree(v)\n",
    "        # 1) Best coverage (fraction of cluster captured)\n",
    "        if cov > best_cov or (cov == best_cov and depth[v] > best_cov_depth):\n",
    "            best_cov, best_cov_node, best_cov_depth = cov, v, depth[v]\n",
    "\n",
    "        # 2) Best Jaccard (penalizes very big clades)\n",
    "        j = cov / (len(C) + subtree_size[v] - cov)\n",
    "        if j > best_j or (j == best_j and depth[v] > best_j_depth):\n",
    "            best_j, best_j_node, best_j_depth = j, v, depth[v]\n",
    "\n",
    "        # 3) α-majority clade (deepest node capturing >= α of the cluster)\n",
    "        if cov >= thr and depth[v] > alpha_depth:\n",
    "            alpha_node, alpha_cov, alpha_depth = v, cov, depth[v]\n",
    "\n",
    "    result = {\n",
    "        \"n_in_cluster\": len(C),\n",
    "        \"n_skipped\": skipped,\n",
    "        \"best_coverage_node\": best_cov_node,\n",
    "        \"best_coverage_fraction\": best_cov / len(C),\n",
    "        \"best_coverage_subtree_size\": subtree_size[best_cov_node] if best_cov_node else None,\n",
    "        \"best_jaccard_node\": best_j_node,\n",
    "        \"best_jaccard\": best_j,\n",
    "        \"best_jaccard_subtree_size\": subtree_size[best_j_node] if best_j_node else None,\n",
    "        \"alpha\": alpha,\n",
    "        \"alpha_node\": alpha_node,\n",
    "        \"alpha_fraction\": (alpha_cov / len(C)) if alpha_node else 0.0,\n",
    "        \"alpha_subtree_size\": subtree_size[alpha_node] if alpha_node else None,\n",
    "        \"alpha_outliers\": (len(C) - alpha_cov) if alpha_node else len(C) - best_cov,\n",
    "    }\n",
    "    return result\n",
    "\n",
    "def cluster_core_and_outliers(idx, cluster_nodes, chosen_node):\n",
    "    \"\"\"Split a cluster into the core inside chosen_node's subtree and the outliers.\"\"\"\n",
    "    if chosen_node is None:\n",
    "        return [], list(cluster_nodes)\n",
    "    tin, tout = idx[\"tin\"], idx[\"tout\"]\n",
    "    a, b = tin[chosen_node], tout[chosen_node]\n",
    "    core, out = [], []\n",
    "    for u in cluster_nodes:\n",
    "        u2 = _normalize(u)\n",
    "        if u2 in tin and a <= tin[u2] <= b:\n",
    "            core.append(u)\n",
    "        else:\n",
    "            out.append(u)\n",
    "    return core, out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "f4261713",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx = index_tree(tree_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "27e071b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster 1/48\n",
      "  in_tree=42  skipped=0\n",
      "  Majority α=0.75: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best coverage: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best Jaccard: 64898 [Aquificaceae]  J=0.033  size=114\n",
      "  Representative clade: 3379134 [Pseudomonadati]\n",
      "  Core=42  Outliers=0\n",
      "    n_in_cluster: 42\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 3379134\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 290974\n",
      "    best_jaccard_node: 64898\n",
      "    best_jaccard: 0.033112582781456956\n",
      "    best_jaccard_subtree_size: 114\n",
      "    alpha: 0.75\n",
      "    alpha_node: 3379134\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 290974\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 2/48\n",
      "  in_tree=31  skipped=0\n",
      "  Majority α=0.75: 356 [Hyphomicrobiales]  frac=0.871  size=36655\n",
      "  Best coverage: 28211 [Alphaproteobacteria]  frac=1.000  size=71591\n",
      "  Best Jaccard: 176299 [Agrobacterium fabrum str. C58]  J=0.032  size=1\n",
      "  Representative clade: 356 [Hyphomicrobiales]\n",
      "  Core=27  Outliers=4\n",
      "    n_in_cluster: 31\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 28211\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 71591\n",
      "    best_jaccard_node: 176299\n",
      "    best_jaccard: 0.03225806451612903\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 356\n",
      "    alpha_fraction: 0.8709677419354839\n",
      "    alpha_subtree_size: 36655\n",
      "    alpha_outliers: 4\n",
      "\n",
      "Cluster 3/48\n",
      "  in_tree=40  skipped=0\n",
      "  Majority α=0.75: 186801 [Clostridia]  frac=0.975  size=13369\n",
      "  Best coverage: 1239 [Bacillota]  frac=1.000  size=94865\n",
      "  Best Jaccard: 478749 [Marvinbryantia formatexigens DSM 14469]  J=0.025  size=1\n",
      "  Representative clade: 186801 [Clostridia]\n",
      "  Core=39  Outliers=1\n",
      "    n_in_cluster: 40\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1239\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 94865\n",
      "    best_jaccard_node: 478749\n",
      "    best_jaccard: 0.025\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 186801\n",
      "    alpha_fraction: 0.975\n",
      "    alpha_subtree_size: 13369\n",
      "    alpha_outliers: 1\n",
      "\n",
      "Cluster 4/48\n",
      "  in_tree=22  skipped=0\n",
      "  Majority α=0.75: 3028117 [Cyanophyceae]  frac=1.000  size=34280\n",
      "  Best coverage: 3028117 [Cyanophyceae]  frac=1.000  size=34280\n",
      "  Best Jaccard: 1111708 [Synechocystis sp. PCC 6803 substr. Kazusa]  J=0.045  size=1\n",
      "  Representative clade: 3028117 [Cyanophyceae]\n",
      "  Core=22  Outliers=0\n",
      "    n_in_cluster: 22\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 3028117\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 34280\n",
      "    best_jaccard_node: 1111708\n",
      "    best_jaccard: 0.045454545454545456\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 3028117\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 34280\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 5/48\n",
      "  in_tree=52  skipped=0\n",
      "  Majority α=0.75: 171549 [Bacteroidales]  frac=0.962  size=4465\n",
      "  Best coverage: 976 [Bacteroidota]  frac=1.000  size=25950\n",
      "  Best Jaccard: 2974251 [Segatella]  J=0.077  size=46\n",
      "  Representative clade: 171549 [Bacteroidales]\n",
      "  Core=50  Outliers=2\n",
      "    n_in_cluster: 52\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 976\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 25950\n",
      "    best_jaccard_node: 2974251\n",
      "    best_jaccard: 0.07692307692307693\n",
      "    best_jaccard_subtree_size: 46\n",
      "    alpha: 0.75\n",
      "    alpha_node: 171549\n",
      "    alpha_fraction: 0.9615384615384616\n",
      "    alpha_subtree_size: 4465\n",
      "    alpha_outliers: 2\n",
      "\n",
      "Cluster 6/48\n",
      "  in_tree=35  skipped=0\n",
      "  Majority α=0.75: 85007 [Mycobacteriales]  frac=0.886  size=15482\n",
      "  Best coverage: 201174 [Actinomycetota]  frac=1.000  size=96386\n",
      "  Best Jaccard: 286801 [Segniliparus]  J=0.048  size=9\n",
      "  Representative clade: 85007 [Mycobacteriales]\n",
      "  Core=31  Outliers=4\n",
      "    n_in_cluster: 35\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 201174\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 96386\n",
      "    best_jaccard_node: 286801\n",
      "    best_jaccard: 0.047619047619047616\n",
      "    best_jaccard_subtree_size: 9\n",
      "    alpha: 0.75\n",
      "    alpha_node: 85007\n",
      "    alpha_fraction: 0.8857142857142857\n",
      "    alpha_subtree_size: 15482\n",
      "    alpha_outliers: 4\n",
      "\n",
      "Cluster 7/48\n",
      "  in_tree=13  skipped=0\n",
      "  Majority α=0.75: 1236 [Gammaproteobacteria]  frac=0.769  size=137913\n",
      "  Best coverage: 1224 [Pseudomonadota]  frac=1.000  size=241984\n",
      "  Best Jaccard: 871585 [Acinetobacter pittii PHEA-2]  J=0.077  size=1\n",
      "  Representative clade: 1236 [Gammaproteobacteria]\n",
      "  Core=10  Outliers=3\n",
      "    n_in_cluster: 13\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1224\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 241984\n",
      "    best_jaccard_node: 871585\n",
      "    best_jaccard: 0.07692307692307693\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1236\n",
      "    alpha_fraction: 0.7692307692307693\n",
      "    alpha_subtree_size: 137913\n",
      "    alpha_outliers: 3\n",
      "\n",
      "Cluster 8/48\n",
      "  in_tree=39  skipped=0\n",
      "  Majority α=0.75: 1760 [Actinomycetes]  frac=0.897  size=93756\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 85018 [Dermatophilaceae]  J=0.029  size=69\n",
      "  Representative clade: 1760 [Actinomycetes]\n",
      "  Core=35  Outliers=4\n",
      "    n_in_cluster: 39\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 85018\n",
      "    best_jaccard: 0.02857142857142857\n",
      "    best_jaccard_subtree_size: 69\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1760\n",
      "    alpha_fraction: 0.8974358974358975\n",
      "    alpha_subtree_size: 93756\n",
      "    alpha_outliers: 4\n",
      "\n",
      "Cluster 9/48\n",
      "  in_tree=13  skipped=0\n",
      "  Majority α=0.75: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 1293054 [Halanaerobium saccharolyticum subsp. saccharolyticum DSM 6643]  J=0.077  size=1\n",
      "  Representative clade: 2 [Bacteria]\n",
      "  Core=13  Outliers=0\n",
      "    n_in_cluster: 13\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 1293054\n",
      "    best_jaccard: 0.07692307692307693\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 2\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 594260\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 10/48\n",
      "  in_tree=20  skipped=0\n",
      "  Majority α=0.75: 1236 [Gammaproteobacteria]  frac=0.800  size=137913\n",
      "  Best coverage: 1224 [Pseudomonadota]  frac=1.000  size=241984\n",
      "  Best Jaccard: 391936 [Isoalcanivorax pacificus W11-5]  J=0.050  size=1\n",
      "  Representative clade: 1236 [Gammaproteobacteria]\n",
      "  Core=16  Outliers=4\n",
      "    n_in_cluster: 20\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1224\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 241984\n",
      "    best_jaccard_node: 391936\n",
      "    best_jaccard: 0.05\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1236\n",
      "    alpha_fraction: 0.8\n",
      "    alpha_subtree_size: 137913\n",
      "    alpha_outliers: 4\n",
      "\n",
      "Cluster 11/48\n",
      "  in_tree=13  skipped=0\n",
      "  Majority α=0.75: 186802 [Eubacteriales]  frac=0.769  size=7979\n",
      "  Best coverage: 186801 [Clostridia]  frac=1.000  size=13369\n",
      "  Best Jaccard: 1121335 [Thermoclostridium stercorarium subsp. stercorarium DSM 8532]  J=0.077  size=1\n",
      "  Representative clade: 186802 [Eubacteriales]\n",
      "  Core=10  Outliers=3\n",
      "    n_in_cluster: 13\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 186801\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 13369\n",
      "    best_jaccard_node: 1121335\n",
      "    best_jaccard: 0.07692307692307693\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 186802\n",
      "    alpha_fraction: 0.7692307692307693\n",
      "    alpha_subtree_size: 7979\n",
      "    alpha_outliers: 3\n",
      "\n",
      "Cluster 12/48\n",
      "  in_tree=46  skipped=0\n",
      "  Majority α=0.75: 1224 [Pseudomonadota]  frac=1.000  size=241984\n",
      "  Best coverage: 1224 [Pseudomonadota]  frac=1.000  size=241984\n",
      "  Best Jaccard: 1227 [Nitrosococcus]  J=0.033  size=16\n",
      "  Representative clade: 1224 [Pseudomonadota]\n",
      "  Core=46  Outliers=0\n",
      "    n_in_cluster: 46\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1224\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 241984\n",
      "    best_jaccard_node: 1227\n",
      "    best_jaccard: 0.03333333333333333\n",
      "    best_jaccard_subtree_size: 16\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1224\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 241984\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 13/48\n",
      "  in_tree=34  skipped=0\n",
      "  Majority α=0.75: 1224 [Pseudomonadota]  frac=1.000  size=241984\n",
      "  Best coverage: 1224 [Pseudomonadota]  frac=1.000  size=241984\n",
      "  Best Jaccard: 155892 [Caulobacter vibrioides]  J=0.056  size=4\n",
      "  Representative clade: 1224 [Pseudomonadota]\n",
      "  Core=34  Outliers=0\n",
      "    n_in_cluster: 34\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1224\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 241984\n",
      "    best_jaccard_node: 155892\n",
      "    best_jaccard: 0.05555555555555555\n",
      "    best_jaccard_subtree_size: 4\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1224\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 241984\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 14/48\n",
      "  in_tree=40  skipped=0\n",
      "  Majority α=0.75: 1236 [Gammaproteobacteria]  frac=1.000  size=137913\n",
      "  Best coverage: 1236 [Gammaproteobacteria]  frac=1.000  size=137913\n",
      "  Best Jaccard: 1621534 [Paraglaciecola]  J=0.029  size=32\n",
      "  Representative clade: 1236 [Gammaproteobacteria]\n",
      "  Core=40  Outliers=0\n",
      "    n_in_cluster: 40\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1236\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 137913\n",
      "    best_jaccard_node: 1621534\n",
      "    best_jaccard: 0.02857142857142857\n",
      "    best_jaccard_subtree_size: 32\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1236\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 137913\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 15/48\n",
      "  in_tree=12  skipped=0\n",
      "  Majority α=0.75: 186822 [Paenibacillaceae]  frac=0.917  size=6575\n",
      "  Best coverage: 1385 [Bacillales]  frac=1.000  size=62961\n",
      "  Best Jaccard: 621372 [Paenibacillus sp. oral taxon 786 str. D14]  J=0.083  size=1\n",
      "  Representative clade: 186822 [Paenibacillaceae]\n",
      "  Core=11  Outliers=1\n",
      "    n_in_cluster: 12\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1385\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 62961\n",
      "    best_jaccard_node: 621372\n",
      "    best_jaccard: 0.08333333333333333\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 186822\n",
      "    alpha_fraction: 0.9166666666666666\n",
      "    alpha_subtree_size: 6575\n",
      "    alpha_outliers: 1\n",
      "\n",
      "Cluster 16/48\n",
      "  in_tree=48  skipped=0\n",
      "  Majority α=0.75: 1760 [Actinomycetes]  frac=1.000  size=93756\n",
      "  Best coverage: 1760 [Actinomycetes]  frac=1.000  size=93756\n",
      "  Best Jaccard: 2838335 [Streptomyces aurantiacus group]  J=0.031  size=19\n",
      "  Representative clade: 1760 [Actinomycetes]\n",
      "  Core=48  Outliers=0\n",
      "    n_in_cluster: 48\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1760\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 93756\n",
      "    best_jaccard_node: 2838335\n",
      "    best_jaccard: 0.03076923076923077\n",
      "    best_jaccard_subtree_size: 19\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1760\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 93756\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 17/48\n",
      "  in_tree=72  skipped=0\n",
      "  Majority α=0.75: 976 [Bacteroidota]  frac=0.986  size=25950\n",
      "  Best coverage: 68336 [Bacteroidota/Chlorobiota group]  frac=1.000  size=26794\n",
      "  Best Jaccard: 379068 [Galbibacter]  J=0.022  size=20\n",
      "  Representative clade: 976 [Bacteroidota]\n",
      "  Core=71  Outliers=1\n",
      "    n_in_cluster: 72\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 68336\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 26794\n",
      "    best_jaccard_node: 379068\n",
      "    best_jaccard: 0.022222222222222223\n",
      "    best_jaccard_subtree_size: 20\n",
      "    alpha: 0.75\n",
      "    alpha_node: 976\n",
      "    alpha_fraction: 0.9861111111111112\n",
      "    alpha_subtree_size: 25950\n",
      "    alpha_outliers: 1\n",
      "\n",
      "Cluster 18/48\n",
      "  in_tree=42  skipped=0\n",
      "  Majority α=0.75: 1385 [Bacillales]  frac=0.976  size=62961\n",
      "  Best coverage: 1239 [Bacillota]  frac=1.000  size=94865\n",
      "  Best Jaccard: 2893056 [Halalkalibacter]  J=0.032  size=23\n",
      "  Representative clade: 1385 [Bacillales]\n",
      "  Core=41  Outliers=1\n",
      "    n_in_cluster: 42\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1239\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 94865\n",
      "    best_jaccard_node: 2893056\n",
      "    best_jaccard: 0.031746031746031744\n",
      "    best_jaccard_subtree_size: 23\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1385\n",
      "    alpha_fraction: 0.9761904761904762\n",
      "    alpha_subtree_size: 62961\n",
      "    alpha_outliers: 1\n",
      "\n",
      "Cluster 19/48\n",
      "  in_tree=29  skipped=0\n",
      "  Majority α=0.75: 3379134 [Pseudomonadati]  frac=0.966  size=290974\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 272561 [Chlamydia trachomatis D/UW-3/CX]  J=0.034  size=1\n",
      "  Representative clade: 3379134 [Pseudomonadati]\n",
      "  Core=28  Outliers=1\n",
      "    n_in_cluster: 29\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 272561\n",
      "    best_jaccard: 0.034482758620689655\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 3379134\n",
      "    alpha_fraction: 0.9655172413793104\n",
      "    alpha_subtree_size: 290974\n",
      "    alpha_outliers: 1\n",
      "\n",
      "Cluster 20/48\n",
      "  in_tree=23  skipped=1\n",
      "  Majority α=0.75: 80840 [Burkholderiales]  frac=1.000  size=23982\n",
      "  Best coverage: 80840 [Burkholderiales]  frac=1.000  size=23982\n",
      "  Best Jaccard: 395019 [Burkholderia multivorans ATCC 17616]  J=0.043  size=1\n",
      "  Representative clade: 80840 [Burkholderiales]\n",
      "  Core=23  Outliers=1\n",
      "    n_in_cluster: 23\n",
      "    n_skipped: 1\n",
      "    best_coverage_node: 80840\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 23982\n",
      "    best_jaccard_node: 395019\n",
      "    best_jaccard: 0.043478260869565216\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 80840\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 23982\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 21/48\n",
      "  in_tree=39  skipped=0\n",
      "  Majority α=0.75: 3028117 [Cyanophyceae]  frac=1.000  size=34280\n",
      "  Best coverage: 3028117 [Cyanophyceae]  frac=1.000  size=34280\n",
      "  Best Jaccard: 33071 [Gloeobacter]  J=0.033  size=24\n",
      "  Representative clade: 3028117 [Cyanophyceae]\n",
      "  Core=39  Outliers=0\n",
      "    n_in_cluster: 39\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 3028117\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 34280\n",
      "    best_jaccard_node: 33071\n",
      "    best_jaccard: 0.03278688524590164\n",
      "    best_jaccard_subtree_size: 24\n",
      "    alpha: 0.75\n",
      "    alpha_node: 3028117\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 34280\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 22/48\n",
      "  in_tree=4  skipped=0\n",
      "  Majority α=0.75: 3031711 [Polyangia]  frac=0.750  size=311\n",
      "  Best coverage: 2818505 [Myxococcota]  frac=1.000  size=1107\n",
      "  Best Jaccard: 246197 [Myxococcus xanthus DK 1622]  J=0.250  size=1\n",
      "  Representative clade: 3031711 [Polyangia]\n",
      "  Core=3  Outliers=1\n",
      "    n_in_cluster: 4\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2818505\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 1107\n",
      "    best_jaccard_node: 246197\n",
      "    best_jaccard: 0.25\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 3031711\n",
      "    alpha_fraction: 0.75\n",
      "    alpha_subtree_size: 311\n",
      "    alpha_outliers: 1\n",
      "\n",
      "Cluster 23/48\n",
      "  in_tree=45  skipped=0\n",
      "  Majority α=0.75: 28211 [Alphaproteobacteria]  frac=1.000  size=71591\n",
      "  Best coverage: 28211 [Alphaproteobacteria]  frac=1.000  size=71591\n",
      "  Best Jaccard: 359391 [Brucella abortus 2308]  J=0.022  size=1\n",
      "  Representative clade: 28211 [Alphaproteobacteria]\n",
      "  Core=45  Outliers=0\n",
      "    n_in_cluster: 45\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 28211\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 71591\n",
      "    best_jaccard_node: 359391\n",
      "    best_jaccard: 0.022222222222222223\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 28211\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 71591\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 24/48\n",
      "  in_tree=10  skipped=0\n",
      "  Majority α=0.75: 203683 [Planctomycetia]  frac=0.800  size=1101\n",
      "  Best coverage: 1783257 [PVC group]  frac=1.000  size=4702\n",
      "  Best Jaccard: 530564 [Pirellula staleyi DSM 6068]  J=0.100  size=1\n",
      "  Representative clade: 203683 [Planctomycetia]\n",
      "  Core=8  Outliers=2\n",
      "    n_in_cluster: 10\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1783257\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 4702\n",
      "    best_jaccard_node: 530564\n",
      "    best_jaccard: 0.1\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 203683\n",
      "    alpha_fraction: 0.8\n",
      "    alpha_subtree_size: 1101\n",
      "    alpha_outliers: 2\n",
      "\n",
      "Cluster 25/48\n",
      "  in_tree=36  skipped=0\n",
      "  Majority α=0.75: 544448 [Mycoplasmatota]  frac=1.000  size=4258\n",
      "  Best coverage: 544448 [Mycoplasmatota]  frac=1.000  size=4258\n",
      "  Best Jaccard: 2895623 [Metamycoplasmataceae]  J=0.075  size=180\n",
      "  Representative clade: 544448 [Mycoplasmatota]\n",
      "  Core=36  Outliers=0\n",
      "    n_in_cluster: 36\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 544448\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 4258\n",
      "    best_jaccard_node: 2895623\n",
      "    best_jaccard: 0.07462686567164178\n",
      "    best_jaccard_subtree_size: 180\n",
      "    alpha: 0.75\n",
      "    alpha_node: 544448\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 4258\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 26/48\n",
      "  in_tree=18  skipped=0\n",
      "  Majority α=0.75: 28216 [Betaproteobacteria]  frac=0.889  size=31013\n",
      "  Best coverage: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best Jaccard: 1088540 [Leptospira weilii serovar Topaz str. LT2116]  J=0.056  size=1\n",
      "  Representative clade: 28216 [Betaproteobacteria]\n",
      "  Core=16  Outliers=2\n",
      "    n_in_cluster: 18\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 3379134\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 290974\n",
      "    best_jaccard_node: 1088540\n",
      "    best_jaccard: 0.05555555555555555\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 28216\n",
      "    alpha_fraction: 0.8888888888888888\n",
      "    alpha_subtree_size: 31013\n",
      "    alpha_outliers: 2\n",
      "\n",
      "Cluster 27/48\n",
      "  in_tree=11  skipped=0\n",
      "  Majority α=0.75: 3379134 [Pseudomonadati]  frac=0.909  size=290974\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 309807 [Salinibacter ruber DSM 13855]  J=0.091  size=1\n",
      "  Representative clade: 3379134 [Pseudomonadati]\n",
      "  Core=10  Outliers=1\n",
      "    n_in_cluster: 11\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 309807\n",
      "    best_jaccard: 0.09090909090909091\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 3379134\n",
      "    alpha_fraction: 0.9090909090909091\n",
      "    alpha_subtree_size: 290974\n",
      "    alpha_outliers: 1\n",
      "\n",
      "Cluster 28/48\n",
      "  in_tree=48  skipped=0\n",
      "  Majority α=0.75: 200940 [Thermodesulfobacteriota]  frac=0.833  size=3185\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 2910589 [Geotalea]  J=0.034  size=12\n",
      "  Representative clade: 200940 [Thermodesulfobacteriota]\n",
      "  Core=40  Outliers=8\n",
      "    n_in_cluster: 48\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 2910589\n",
      "    best_jaccard: 0.034482758620689655\n",
      "    best_jaccard_subtree_size: 12\n",
      "    alpha: 0.75\n",
      "    alpha_node: 200940\n",
      "    alpha_fraction: 0.8333333333333334\n",
      "    alpha_subtree_size: 3185\n",
      "    alpha_outliers: 8\n",
      "\n",
      "Cluster 29/48\n",
      "  in_tree=14  skipped=0\n",
      "  Majority α=0.75: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best coverage: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best Jaccard: 1099 [Pelodictyon]  J=0.091  size=10\n",
      "  Representative clade: 3379134 [Pseudomonadati]\n",
      "  Core=14  Outliers=0\n",
      "    n_in_cluster: 14\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 3379134\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 290974\n",
      "    best_jaccard_node: 1099\n",
      "    best_jaccard: 0.09090909090909091\n",
      "    best_jaccard_subtree_size: 10\n",
      "    alpha: 0.75\n",
      "    alpha_node: 3379134\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 290974\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 30/48\n",
      "  in_tree=7  skipped=0\n",
      "  Majority α=0.75: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best coverage: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best Jaccard: 320771 [Pedosphaera parvula Ellin514]  J=0.143  size=1\n",
      "  Representative clade: 3379134 [Pseudomonadati]\n",
      "  Core=7  Outliers=0\n",
      "    n_in_cluster: 7\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 3379134\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 290974\n",
      "    best_jaccard_node: 320771\n",
      "    best_jaccard: 0.14285714285714285\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 3379134\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 290974\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 31/48\n",
      "  in_tree=35  skipped=0\n",
      "  Majority α=0.75: 1239 [Bacillota]  frac=0.771  size=94865\n",
      "  Best coverage: 1783272 [Bacillati]  frac=1.000  size=233820\n",
      "  Best Jaccard: 2916693 [Desulforamulus]  J=0.061  size=17\n",
      "  Representative clade: 1239 [Bacillota]\n",
      "  Core=27  Outliers=8\n",
      "    n_in_cluster: 35\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1783272\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 233820\n",
      "    best_jaccard_node: 2916693\n",
      "    best_jaccard: 0.061224489795918366\n",
      "    best_jaccard_subtree_size: 17\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1239\n",
      "    alpha_fraction: 0.7714285714285715\n",
      "    alpha_subtree_size: 94865\n",
      "    alpha_outliers: 8\n",
      "\n",
      "Cluster 32/48\n",
      "  in_tree=27  skipped=0\n",
      "  Majority α=0.75: 28211 [Alphaproteobacteria]  frac=1.000  size=71591\n",
      "  Best coverage: 28211 [Alphaproteobacteria]  frac=1.000  size=71591\n",
      "  Best Jaccard: 264203 [Zymomonas mobilis subsp. mobilis ZM4 = ATCC 31821]  J=0.037  size=1\n",
      "  Representative clade: 28211 [Alphaproteobacteria]\n",
      "  Core=27  Outliers=0\n",
      "    n_in_cluster: 27\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 28211\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 71591\n",
      "    best_jaccard_node: 264203\n",
      "    best_jaccard: 0.037037037037037035\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 28211\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 71591\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 33/48\n",
      "  in_tree=20  skipped=0\n",
      "  Majority α=0.75: 136 [Spirochaetales]  frac=0.900  size=2082\n",
      "  Best coverage: 203692 [Spirochaetia]  frac=1.000  size=3782\n",
      "  Best Jaccard: 2951104 [Breznakiellaceae]  J=0.056  size=18\n",
      "  Representative clade: 136 [Spirochaetales]\n",
      "  Core=18  Outliers=2\n",
      "    n_in_cluster: 20\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 203692\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 3782\n",
      "    best_jaccard_node: 2951104\n",
      "    best_jaccard: 0.05555555555555555\n",
      "    best_jaccard_subtree_size: 18\n",
      "    alpha: 0.75\n",
      "    alpha_node: 136\n",
      "    alpha_fraction: 0.9\n",
      "    alpha_subtree_size: 2082\n",
      "    alpha_outliers: 2\n",
      "\n",
      "Cluster 34/48\n",
      "  in_tree=40  skipped=0\n",
      "  Majority α=0.75: 91347 [Enterobacterales]  frac=1.000  size=39984\n",
      "  Best coverage: 91347 [Enterobacterales]  frac=1.000  size=39984\n",
      "  Best Jaccard: 1125630 [Klebsiella pneumoniae subsp. pneumoniae HS11286]  J=0.025  size=1\n",
      "  Representative clade: 91347 [Enterobacterales]\n",
      "  Core=40  Outliers=0\n",
      "    n_in_cluster: 40\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 91347\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 39984\n",
      "    best_jaccard_node: 1125630\n",
      "    best_jaccard: 0.025\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 91347\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 39984\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 35/48\n",
      "  in_tree=32  skipped=0\n",
      "  Majority α=0.75: 186801 [Clostridia]  frac=0.750  size=13369\n",
      "  Best coverage: 1783272 [Bacillati]  frac=1.000  size=233820\n",
      "  Best Jaccard: 441771 [Clostridium botulinum A str. Hall]  J=0.031  size=1\n",
      "  Representative clade: 186801 [Clostridia]\n",
      "  Core=24  Outliers=8\n",
      "    n_in_cluster: 32\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1783272\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 233820\n",
      "    best_jaccard_node: 441771\n",
      "    best_jaccard: 0.03125\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 186801\n",
      "    alpha_fraction: 0.75\n",
      "    alpha_subtree_size: 13369\n",
      "    alpha_outliers: 8\n",
      "\n",
      "Cluster 36/48\n",
      "  in_tree=57  skipped=0\n",
      "  Majority α=0.75: 1760 [Actinomycetes]  frac=1.000  size=93756\n",
      "  Best coverage: 1760 [Actinomycetes]  frac=1.000  size=93756\n",
      "  Best Jaccard: 196081 [Scardovia]  J=0.029  size=13\n",
      "  Representative clade: 1760 [Actinomycetes]\n",
      "  Core=57  Outliers=0\n",
      "    n_in_cluster: 57\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1760\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 93756\n",
      "    best_jaccard_node: 196081\n",
      "    best_jaccard: 0.029411764705882353\n",
      "    best_jaccard_subtree_size: 13\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1760\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 93756\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 37/48\n",
      "  in_tree=15  skipped=0\n",
      "  Majority α=0.75: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 479434 [Sphaerobacter thermophilus DSM 20745]  J=0.067  size=1\n",
      "  Representative clade: 2 [Bacteria]\n",
      "  Core=15  Outliers=0\n",
      "    n_in_cluster: 15\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 479434\n",
      "    best_jaccard: 0.06666666666666667\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 2\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 594260\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 38/48\n",
      "  in_tree=27  skipped=0\n",
      "  Majority α=0.75: 204457 [Sphingomonadales]  frac=0.889  size=8227\n",
      "  Best coverage: 28211 [Alphaproteobacteria]  frac=1.000  size=71591\n",
      "  Best Jaccard: 332055 [Sphingobium indicum]  J=0.065  size=6\n",
      "  Representative clade: 204457 [Sphingomonadales]\n",
      "  Core=24  Outliers=3\n",
      "    n_in_cluster: 27\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 28211\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 71591\n",
      "    best_jaccard_node: 332055\n",
      "    best_jaccard: 0.06451612903225806\n",
      "    best_jaccard_subtree_size: 6\n",
      "    alpha: 0.75\n",
      "    alpha_node: 204457\n",
      "    alpha_fraction: 0.8888888888888888\n",
      "    alpha_subtree_size: 8227\n",
      "    alpha_outliers: 3\n",
      "\n",
      "Cluster 39/48\n",
      "  in_tree=19  skipped=0\n",
      "  Majority α=0.75: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 1321779 [Leptotrichia sp. oral taxon 215 str. W9775]  J=0.053  size=1\n",
      "  Representative clade: 2 [Bacteria]\n",
      "  Core=19  Outliers=0\n",
      "    n_in_cluster: 19\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 1321779\n",
      "    best_jaccard: 0.05263157894736842\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 2\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 594260\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 40/48\n",
      "  in_tree=16  skipped=0\n",
      "  Majority α=0.75: 1239 [Bacillota]  frac=1.000  size=94865\n",
      "  Best coverage: 1239 [Bacillota]  frac=1.000  size=94865\n",
      "  Best Jaccard: 273068 [Caldanaerobacter subterraneus subsp. tengcongensis MB4]  J=0.062  size=1\n",
      "  Representative clade: 1239 [Bacillota]\n",
      "  Core=16  Outliers=0\n",
      "    n_in_cluster: 16\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1239\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 94865\n",
      "    best_jaccard_node: 273068\n",
      "    best_jaccard: 0.0625\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1239\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 94865\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 41/48\n",
      "  in_tree=3  skipped=0\n",
      "  Majority α=0.75: 1783272 [Bacillati]  frac=1.000  size=233820\n",
      "  Best coverage: 1783272 [Bacillati]  frac=1.000  size=233820\n",
      "  Best Jaccard: 679936 [Sulfobacillus acidophilus DSM 10332]  J=0.333  size=1\n",
      "  Representative clade: 1783272 [Bacillati]\n",
      "  Core=3  Outliers=0\n",
      "    n_in_cluster: 3\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1783272\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 233820\n",
      "    best_jaccard_node: 679936\n",
      "    best_jaccard: 0.3333333333333333\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1783272\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 233820\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 42/48\n",
      "  in_tree=78  skipped=0\n",
      "  Majority α=0.75: 186826 [Lactobacillales]  frac=0.936  size=13989\n",
      "  Best coverage: 1783272 [Bacillati]  frac=1.000  size=233820\n",
      "  Best Jaccard: 2767887 [Ligilactobacillus]  J=0.027  size=74\n",
      "  Representative clade: 186826 [Lactobacillales]\n",
      "  Core=73  Outliers=5\n",
      "    n_in_cluster: 78\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1783272\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 233820\n",
      "    best_jaccard_node: 2767887\n",
      "    best_jaccard: 0.02702702702702703\n",
      "    best_jaccard_subtree_size: 74\n",
      "    alpha: 0.75\n",
      "    alpha_node: 186826\n",
      "    alpha_fraction: 0.9358974358974359\n",
      "    alpha_subtree_size: 13989\n",
      "    alpha_outliers: 5\n",
      "\n",
      "Cluster 43/48\n",
      "  in_tree=12  skipped=0\n",
      "  Majority α=0.75: 1236 [Gammaproteobacteria]  frac=0.833  size=137913\n",
      "  Best coverage: 1224 [Pseudomonadota]  frac=1.000  size=241984\n",
      "  Best Jaccard: 223283 [Pseudomonas syringae pv. tomato str. DC3000]  J=0.083  size=1\n",
      "  Representative clade: 1236 [Gammaproteobacteria]\n",
      "  Core=10  Outliers=2\n",
      "    n_in_cluster: 12\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1224\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 241984\n",
      "    best_jaccard_node: 223283\n",
      "    best_jaccard: 0.08333333333333333\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1236\n",
      "    alpha_fraction: 0.8333333333333334\n",
      "    alpha_subtree_size: 137913\n",
      "    alpha_outliers: 2\n",
      "\n",
      "Cluster 44/48\n",
      "  in_tree=8  skipped=0\n",
      "  Majority α=0.75: 204433 [Terriglobales]  frac=0.750  size=421\n",
      "  Best coverage: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best Jaccard: 1129897 [Nitrolancea hollandica Lb]  J=0.125  size=1\n",
      "  Representative clade: 204433 [Terriglobales]\n",
      "  Core=6  Outliers=2\n",
      "    n_in_cluster: 8\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 3379134\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 290974\n",
      "    best_jaccard_node: 1129897\n",
      "    best_jaccard: 0.125\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 204433\n",
      "    alpha_fraction: 0.75\n",
      "    alpha_subtree_size: 421\n",
      "    alpha_outliers: 2\n",
      "\n",
      "Cluster 45/48\n",
      "  in_tree=21  skipped=0\n",
      "  Majority α=0.75: 1783272 [Bacillati]  frac=0.857  size=233820\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 272632 [Mycoplasma mycoides subsp. mycoides SC str. PG1]  J=0.048  size=1\n",
      "  Representative clade: 1783272 [Bacillati]\n",
      "  Core=18  Outliers=3\n",
      "    n_in_cluster: 21\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 272632\n",
      "    best_jaccard: 0.047619047619047616\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1783272\n",
      "    alpha_fraction: 0.8571428571428571\n",
      "    alpha_subtree_size: 233820\n",
      "    alpha_outliers: 3\n",
      "\n",
      "Cluster 46/48\n",
      "  in_tree=28  skipped=0\n",
      "  Majority α=0.75: 1239 [Bacillota]  frac=0.964  size=94865\n",
      "  Best coverage: 1783272 [Bacillati]  frac=1.000  size=233820\n",
      "  Best Jaccard: 3106714 [unclassified Anaerovoracaceae Incertae Sedis]  J=0.085  size=23\n",
      "  Representative clade: 1239 [Bacillota]\n",
      "  Core=27  Outliers=1\n",
      "    n_in_cluster: 28\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 1783272\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 233820\n",
      "    best_jaccard_node: 3106714\n",
      "    best_jaccard: 0.0851063829787234\n",
      "    best_jaccard_subtree_size: 23\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1239\n",
      "    alpha_fraction: 0.9642857142857143\n",
      "    alpha_subtree_size: 94865\n",
      "    alpha_outliers: 1\n",
      "\n",
      "Cluster 47/48\n",
      "  in_tree=11  skipped=0\n",
      "  Majority α=0.75: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best coverage: 2 [Bacteria]  frac=1.000  size=594260\n",
      "  Best Jaccard: 511051 [Caldisericum exile AZM16c01]  J=0.091  size=1\n",
      "  Representative clade: 2 [Bacteria]\n",
      "  Core=11  Outliers=0\n",
      "    n_in_cluster: 11\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 2\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 594260\n",
      "    best_jaccard_node: 511051\n",
      "    best_jaccard: 0.09090909090909091\n",
      "    best_jaccard_subtree_size: 1\n",
      "    alpha: 0.75\n",
      "    alpha_node: 2\n",
      "    alpha_fraction: 1.0\n",
      "    alpha_subtree_size: 594260\n",
      "    alpha_outliers: 0\n",
      "\n",
      "Cluster 48/48\n",
      "  in_tree=54  skipped=0\n",
      "  Majority α=0.75: 1224 [Pseudomonadota]  frac=0.963  size=241984\n",
      "  Best coverage: 3379134 [Pseudomonadati]  frac=1.000  size=290974\n",
      "  Best Jaccard: 138073 [Candidatus Regiella insecticola]  J=0.036  size=3\n",
      "  Representative clade: 1224 [Pseudomonadota]\n",
      "  Core=52  Outliers=2\n",
      "    n_in_cluster: 54\n",
      "    n_skipped: 0\n",
      "    best_coverage_node: 3379134\n",
      "    best_coverage_fraction: 1.0\n",
      "    best_coverage_subtree_size: 290974\n",
      "    best_jaccard_node: 138073\n",
      "    best_jaccard: 0.03636363636363636\n",
      "    best_jaccard_subtree_size: 3\n",
      "    alpha: 0.75\n",
      "    alpha_node: 1224\n",
      "    alpha_fraction: 0.9629629629629629\n",
      "    alpha_subtree_size: 241984\n",
      "    alpha_outliers: 2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def fmt_node(node, tid2nm):\n",
    "    \"\"\"Format a node as 'taxid [name]' if name exists; handle None.\"\"\"\n",
    "    if node is None:\n",
    "        return \"None\"\n",
    "    name = tid2nm.get(node)\n",
    "    return f\"{node} [{name}]\" if name else str(node)\n",
    "\n",
    "def summarize_clusters(idx, clusters, tid2nm, alpha=0.75, size_cap=None, show_all=False):\n",
    "    \"\"\"\n",
    "    Print a compact summary per cluster with translated node labels in brackets.\n",
    "    Set show_all=True to also dump the full stats dict.\n",
    "    \"\"\"\n",
    "    n_clusters = len(clusters)\n",
    "    for i, C in enumerate(clusters, 1):\n",
    "        stats = cluster_tree_stats(idx, C, alpha=alpha, size_cap=size_cap)\n",
    "        rep = stats.get(\"alpha_node\") or stats.get(\"best_jaccard_node\")\n",
    "        core, outliers = cluster_core_and_outliers(idx, C, rep)\n",
    "\n",
    "        ac = stats.get  # tiny shortcut\n",
    "\n",
    "        print(f\"Cluster {i}/{n_clusters}\")\n",
    "        print(f\"  in_tree={ac('n_in_cluster',0)}  skipped={ac('n_skipped',0)}\")\n",
    "\n",
    "        # Majority (alpha) clade\n",
    "        print(f\"  Majority α={ac('alpha')}: {fmt_node(ac('alpha_node'), tid2nm)}  \"\n",
    "              f\"frac={ac('alpha_fraction',0):.3f}  size={ac('alpha_subtree_size')}\")\n",
    "\n",
    "        # Best coverage clade\n",
    "        print(f\"  Best coverage: {fmt_node(ac('best_coverage_node'), tid2nm)}  \"\n",
    "              f\"frac={ac('best_coverage_fraction',0):.3f}  \"\n",
    "              f\"size={ac('best_coverage_subtree_size')}\")\n",
    "\n",
    "        # Best Jaccard clade\n",
    "        print(f\"  Best Jaccard: {fmt_node(ac('best_jaccard_node'), tid2nm)}  \"\n",
    "              f\"J={ac('best_jaccard',0):.3f}  size={ac('best_jaccard_subtree_size')}\")\n",
    "\n",
    "        # Representative clade (alpha if present, else best Jaccard)\n",
    "        print(f\"  Representative clade: {fmt_node(rep, tid2nm)}\")\n",
    "\n",
    "        print(f\"  Core={len(core)}  Outliers={len(outliers)}\")\n",
    "        if show_all:\n",
    "            for k, v in stats.items():\n",
    "                print(f\"    {k}: {v}\")\n",
    "        print()\n",
    "        \n",
    "summarize_clusters(idx, clusters.values(), tid2nm, alpha=0.75, size_cap=None, show_all=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "2cda71dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_separation(idx, reps):\n",
    "    \"\"\"\n",
    "    Compute how distinct clusters' representative clades are on the tree.\n",
    "\n",
    "    Returns:\n",
    "      - overlap_matrix[i,j] = True if subtrees overlap (nested or intersect)\n",
    "      - distance_matrix[i,j] = edge distance between representative nodes\n",
    "      - separation_fraction = fraction of cluster pairs that are disjoint\n",
    "      - mean_interclade_distance = mean distance across distinct clades\n",
    "    \"\"\"\n",
    "    tin, tout, depth, parent = idx[\"tin\"], idx[\"tout\"], idx[\"depth\"], idx[\"parent\"]\n",
    "    n = len(reps)\n",
    "    overlap = np.zeros((n, n), dtype=bool)\n",
    "    dist = np.zeros((n, n), dtype=float)\n",
    "\n",
    "    # helper to find LCA distance\n",
    "    def lca_distance(u, v):\n",
    "        if u is None or v is None:\n",
    "            return np.nan\n",
    "        a, b = u, v\n",
    "        while not (tin[a] <= tin[b] <= tout[a]) and not (tin[b] <= tin[a] <= tout[b]):\n",
    "            if depth[a] > depth[b]:\n",
    "                a = parent[a]\n",
    "            else:\n",
    "                b = parent[b]\n",
    "        lca = a if tin[a] <= tin[b] <= tout[a] else b\n",
    "        return depth[u] + depth[v] - 2 * depth[lca]\n",
    "\n",
    "    for i in range(n):\n",
    "        a = reps[i]\n",
    "        if a is None:\n",
    "            continue\n",
    "        for j in range(i + 1, n):\n",
    "            b = reps[j]\n",
    "            if b is None:\n",
    "                continue\n",
    "            # overlap test\n",
    "            if (tin[a] <= tin[b] <= tout[a]) or (tin[b] <= tin[a] <= tout[b]):\n",
    "                overlap[i, j] = overlap[j, i] = True\n",
    "            # distance\n",
    "            dist[i, j] = dist[j, i] = lca_distance(a, b)\n",
    "\n",
    "    # fraction of disjoint (non-overlapping) pairs\n",
    "    nonoverlap_fraction = 1 - overlap[np.triu_indices(n, 1)].mean()\n",
    "    # mean interclade distance ignoring NaN\n",
    "    mean_dist = np.nanmean(dist[np.triu_indices(n, 1)])\n",
    "    return {\n",
    "        \"overlap_matrix\": overlap,\n",
    "        \"distance_matrix\": dist,\n",
    "        \"fraction_disjoint\": nonoverlap_fraction,\n",
    "        \"mean_interclade_distance\": mean_dist,\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "27fabef4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Global cluster separation:\n",
      "  Fraction of disjoint clades: 0.673\n",
      "  Mean interclade distance: 4.67\n",
      "\n",
      "Pairwise overlap (Y=overlap, .=disjoint):\n",
      "Cluster  1: .Y..Y.Y.YY.YYY..Y.YY.YYY.YYYYY.YYY..YYY...YY..YY\n",
      "Cluster  2: Y.......Y..YY.....Y.......Y.Y..Y....Y.Y.......YY\n",
      "Cluster  3: ........Y.Y...................Y...Y.Y.YY....YYY.\n",
      "Cluster  4: ........Y...........Y...............Y.Y.....Y.Y.\n",
      "Cluster  5: Y.......Y.......Y.Y.......Y.Y.......Y.Y.......Y.\n",
      "Cluster  6: .......YY......Y...................YY.Y.....Y.Y.\n",
      "Cluster  7: Y.......YY.YYY....Y.......Y.Y....Y..Y.Y...Y...YY\n",
      "Cluster  8: .....Y..Y......Y...................YY.Y.....Y.Y.\n",
      "Cluster  9: YYYYYYYY.YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY\n",
      "Cluster 10: Y.....Y.Y..YYY....Y.......Y.Y....Y..Y.Y...Y...YY\n",
      "Cluster 11: ..Y.....Y.....................Y...Y.Y.YY....YYY.\n",
      "Cluster 12: YY....Y.YY..YY....YY..Y..YY.Y..Y.Y..YYY...Y...YY\n",
      "Cluster 13: YY....Y.YY.Y.Y....YY..Y..YY.Y..Y.Y..YYY...Y...YY\n",
      "Cluster 14: Y.....Y.YY.YY.....Y.......Y.Y....Y..Y.Y...Y...YY\n",
      "Cluster 15: ........Y........Y..................Y.Y.....YYY.\n",
      "Cluster 16: .....Y.YY..........................YY.Y.....Y.Y.\n",
      "Cluster 17: Y...Y...Y.........Y.......Y.Y.......Y.Y.......Y.\n",
      "Cluster 18: ........Y.....Y.....................Y.Y.....YYY.\n",
      "Cluster 19: YY..Y.Y.YY.YYY..Y..Y.YYY.YYYYY.YYY..YYY...YY..YY\n",
      "Cluster 20: Y.......Y..YY.....Y......YY.Y.......Y.Y.......YY\n",
      "Cluster 21: ...Y....Y...........................Y.Y.....Y.Y.\n",
      "Cluster 22: Y.......Y.........Y.......Y.Y.......Y.Y.......Y.\n",
      "Cluster 23: Y.......Y..YY.....Y.......Y.Y..Y....Y.Y.......YY\n",
      "Cluster 24: Y.......Y.........Y.......Y.Y.......Y.Y.......Y.\n",
      "Cluster 25: ........Y...........................Y.Y.....Y.Y.\n",
      "Cluster 26: Y.......Y..YY.....YY......Y.Y.......Y.Y.......YY\n",
      "Cluster 27: YY..Y.Y.YY.YYY..Y.YY.YYY.Y.YYY.YYY..YYY...YY..YY\n",
      "Cluster 28: Y.......Y.........Y.......Y.Y.......Y.Y.......Y.\n",
      "Cluster 29: YY..Y.Y.YY.YYY..Y.YY.YYY.YYY.Y.YYY..YYY...YY..YY\n",
      "Cluster 30: Y.......Y.........Y.......Y.Y.......Y.Y.......Y.\n",
      "Cluster 31: ..Y.....Y.Y.......................Y.Y.YY....YYY.\n",
      "Cluster 32: YY......Y..YY.....Y...Y...Y.Y.......YYY.......YY\n",
      "Cluster 33: Y.......Y.........Y.......Y.Y.......Y.Y.......Y.\n",
      "Cluster 34: Y.....Y.YY.YYY....Y.......Y.Y.......Y.Y.......YY\n",
      "Cluster 35: ..Y.....Y.Y...................Y.....Y.YY....YYY.\n",
      "Cluster 36: .....Y.YY......Y....................Y.Y.....Y.Y.\n",
      "Cluster 37: YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY.YYYYYYYYYYY\n",
      "Cluster 38: Y.......Y..YY.....Y.......Y.Y..Y....Y.Y.......YY\n",
      "Cluster 39: YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY.YYYYYYYYY\n",
      "Cluster 40: ..Y.....Y.Y...................Y...Y.Y.Y.....YYY.\n",
      "Cluster 41: ........Y...........................Y.Y.....Y.Y.\n",
      "Cluster 42: ........Y...........................Y.Y.....YYY.\n",
      "Cluster 43: Y.....Y.YY.YYY....Y.......Y.Y.......Y.Y.......YY\n",
      "Cluster 44: Y.......Y.........Y.......Y.Y.......Y.Y.......Y.\n",
      "Cluster 45: ..YY.Y.YY.Y...YY.Y..Y...Y.....Y...YYY.YYYY...YY.\n",
      "Cluster 46: ..Y.....Y.Y...Y..Y............Y...Y.Y.YY.Y..Y.Y.\n",
      "Cluster 47: YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY.Y\n",
      "Cluster 48: YY....Y.YY.YYY....YY..Y..YY.Y..Y.Y..YYY...Y...Y.\n"
     ]
    }
   ],
   "source": [
    "def summarize_separation(idx, clusters, tid2nm, alpha=0.65):\n",
    "    \"\"\"Print global separation stats and a compact matrix of overlaps.\"\"\"\n",
    "    reps = []\n",
    "    for C in clusters:\n",
    "        stats = cluster_tree_stats(idx, C, alpha=alpha)\n",
    "        reps.append(stats.get(\"alpha_node\") or stats.get(\"best_jaccard_node\"))\n",
    "\n",
    "    sep = cluster_separation(idx, reps)\n",
    "    print(f\"\\nGlobal cluster separation:\")\n",
    "    print(f\"  Fraction of disjoint clades: {sep['fraction_disjoint']:.3f}\")\n",
    "    print(f\"  Mean interclade distance: {sep['mean_interclade_distance']:.2f}\\n\")\n",
    "\n",
    "    # show which clusters overlap\n",
    "    n = len(reps)\n",
    "    print(\"Pairwise overlap (Y=overlap, .=disjoint):\")\n",
    "    for i in range(n):\n",
    "        row = ''.join('Y' if sep[\"overlap_matrix\"][i,j] else '.' for j in range(n))\n",
    "        print(f\"Cluster {i+1:>2}: {row}\")\n",
    "        \n",
    "summarize_separation(idx, clusters.values(), tid2nm, alpha=0.65)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcf7af69",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lagassoBioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
